{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "wrgOhk8U4Rpl"
      },
      "source": [
        "# Quickstart: Querying PDF With Astra and LangChain\n",
        "\n",
        "### A question-answering demo using Astra DB and LangChain, powered by Vector Search"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MqfJKgRM4Rpo"
      },
      "source": [
        "#### Pre-requisites:\n",
        "\n",
        "You need a **_Serverless Cassandra with Vector Search_** database on [Astra DB](https://astra.datastax.com) to run this demo. As outlined in more detail [here](https://docs.datastax.com/en/astra-serverless/docs/vector-search/quickstart.html#_prepare_for_using_your_vector_database), you should get a DB Token with role _Database Administrator_ and copy your Database ID: these connection parameters are needed momentarily.\n",
        "\n",
        "You also need an [OpenAI API Key](https://cassio.org/start_here/#llm-access) for this demo to work.\n",
        "\n",
        "#### What you will do:\n",
        "\n",
        "- Setup: import dependencies, provide secrets, create the LangChain vector store;\n",
        "- Run a Question-Answering loop retrieving the relevant headlines and having an LLM construct the answer."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "m_FeN-Ep4Rpp"
      },
      "source": [
        "Install the required dependencies:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "Uk0qUhJUQrkO",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "b767952d-9137-4f58-f883-221d6e4533ef"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m41.3/41.3 kB\u001b[0m \u001b[31m968.0 kB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m510.5/510.5 kB\u001b[0m \u001b[31m7.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m817.7/817.7 kB\u001b[0m \u001b[31m14.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m292.8/292.8 kB\u001b[0m \u001b[31m14.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.8/1.8 MB\u001b[0m \u001b[31m20.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m18.9/18.9 MB\u001b[0m \u001b[31m48.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m116.3/116.3 kB\u001b[0m \u001b[31m9.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m194.1/194.1 kB\u001b[0m \u001b[31m9.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m134.8/134.8 kB\u001b[0m \u001b[31m8.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.9/1.9 MB\u001b[0m \u001b[31m36.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m289.1/289.1 kB\u001b[0m \u001b[31m18.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m113.7/113.7 kB\u001b[0m \u001b[31m3.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m75.6/75.6 kB\u001b[0m \u001b[31m6.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m49.4/49.4 kB\u001b[0m \u001b[31m3.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m77.9/77.9 kB\u001b[0m \u001b[31m5.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m58.3/58.3 kB\u001b[0m \u001b[31m5.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m53.0/53.0 kB\u001b[0m \u001b[31m4.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m141.1/141.1 kB\u001b[0m \u001b[31m5.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h"
          ]
        }
      ],
      "source": [
        "!pip install -q cassio datasets langchain openai tiktoken"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XQQN-L2J4Rpq"
      },
      "source": [
        "Import the packages you'll need:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "id": "V4qBIihE4Rpq"
      },
      "outputs": [],
      "source": [
        "# LangChain components to use\n",
        "from langchain.vectorstores.cassandra import Cassandra\n",
        "from langchain.indexes.vectorstore import VectorStoreIndexWrapper\n",
        "from langchain.llms import OpenAI\n",
        "from langchain.embeddings import OpenAIEmbeddings\n",
        "\n",
        "# Support for dataset retrieval with Hugging Face\n",
        "from datasets import load_dataset\n",
        "\n",
        "# With CassIO, the engine powering the Astra DB integration in LangChain,\n",
        "# you will also initialize the DB connection:\n",
        "import cassio"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install PyPDF2"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WIs76OPQ6JyD",
        "outputId": "02f12d40-bc01-4f4e-dcf1-cd939313e4b8"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting PyPDF2\n",
            "  Downloading pypdf2-3.0.1-py3-none-any.whl (232 kB)\n",
            "\u001b[?25l     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.0/232.6 kB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K     \u001b[91m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m\u001b[91m╸\u001b[0m\u001b[90m━\u001b[0m \u001b[32m225.3/232.6 kB\u001b[0m \u001b[31m6.6 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\r\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m232.6/232.6 kB\u001b[0m \u001b[31m5.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: PyPDF2\n",
            "Successfully installed PyPDF2-3.0.1\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from PyPDF2 import PdfReader"
      ],
      "metadata": {
        "id": "1itBNL1v6N9-"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Vu2UauiC4Rpr"
      },
      "source": [
        "### Setup"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {
        "id": "eqpM6I854Rpr"
      },
      "outputs": [],
      "source": [
        "ASTRA_DB_APPLICATION_TOKEN = \"AstraCS:lyokujRudcbbvsEAWziqsaSM:4b109b24afe5c59c8775e8a37f327b8422be30a78f19f01256c20a8f2594d256\" # enter the \"AstraCS:...\" string found in in your Token JSON file\n",
        "ASTRA_DB_ID = \"5d9400f9-7c66-49dc-b310-a4e59c3ccb2b\" # enter your Database ID\n",
        "\n",
        "OPENAI_API_KEY = \"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\" # enter your OpenAI key"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Q1cmD5EF4Rpr"
      },
      "source": [
        "#### Provide your secrets:\n",
        "\n",
        "Replace the following with your Astra DB connection details and your OpenAI API key:"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# provide the path of  pdf file/files.\n",
        "pdfreader = PdfReader('pdfname.pdf')"
      ],
      "metadata": {
        "id": "waVKJW-n6jqJ"
      },
      "execution_count": 27,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from typing_extensions import Concatenate\n",
        "# read text from pdf\n",
        "raw_text = ''\n",
        "for i, page in enumerate(pdfreader.pages):\n",
        "    content = page.extract_text()\n",
        "    if content:\n",
        "        raw_text += content"
      ],
      "metadata": {
        "id": "42BKuFRO6meP"
      },
      "execution_count": 28,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "raw_text"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 140
        },
        "id": "vR41Iq-4ZHnG",
        "outputId": "8143247f-c5de-47d1-ffb6-b8ed392bbd79"
      },
      "execution_count": 29,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'Full Terms & Conditions of access and use can be found at\\nhttps://www.tandfonline.com/action/journalInformation?journalCode=tbsd20\\nJournal of Biomolecular Structure and Dynamics\\nISSN: (Print) (Online) Journal homepage: www.tandfonline.com/journals/tbsd20\\nApplication of deep learning and molecular\\nmodeling to identify small drug-like compounds as\\npotential HIV-1 entry inhibitors\\nAlexander M. Andrianov, Grigory I. Nikolaev, Nikita A. Shuldov, Ivan P. Bosko,\\nArseny I. Anischenko & Alexander V. Tuzikov\\nTo cite this article:  Alexander M. Andrianov, Grigory I. Nikolaev, Nikita A. Shuldov, Ivan P.\\nBosko, Arseny I. Anischenko & Alexander V. Tuzikov (2022) Application of deep learning\\nand molecular modeling to identify small drug-like compounds as potential HIV-1 entry\\ninhibitors, Journal of Biomolecular Structure and Dynamics, 40:16, 7555-7573, DOI:\\n10.1080/07391102.2021.1905559\\nTo link to this article:  https://doi.org/10.1080/07391102.2021.1905559\\nPublished online: 15 Apr 2021.\\nSubmit your article to this journal \\nArticle views: 612\\nView related articles \\nView Crossmark data\\nCiting articles: 7 View citing articles \\nApplication of deep learning and molecular modeling to identify small drug-like\\ncompounds as potential HIV-1 entry inhibitors\\nAlexander M. Andrianova, Grigory I. Nikolaevb, Nikita A. Shuldovc, Ivan P. Boskob, Arseny I. Anischenkocand\\nAlexander V. Tuzikovb\\naInstitute of Bioorganic Chemistry, National Academy of Sciences of Belarus, Minsk, Republic of Belarus;bUnited Institute of Informatics\\nProblems, National Academy of Sciences of Belarus, Minsk, Republic of Belarus;cFaculty of Applied Mathematics & Computer Science,\\nBelarusian State University, Minsk, Republic of Belarus\\nCommunicated by Ramaswamy H. Sarma\\nABSTRACT\\nA generative adversarial autoencoder for the rational design of potential HIV-1 entry inhibitors able to\\nblock CD4-binding site of the viral envelope protein gp120 was developed. To do this, the following\\nstudies were carried out: (i) an autoencoder architecture was constructed; (ii) a virtual compoundlibrary of potential anti-HIV-1 agents for training the neural network was formed by the concept of\\nclick chemistry allowing one to generate a large number of drug candidates by their assembly from\\nsmall modular units; (iii) molecular docking of all compounds from this library with gp120 was made\\nand calculations of the values of binding free energy were performed; (iv) molecular fingerprints of\\nchemical compounds from the training dataset were generated; (v) training of the developed autoen-coder was implemented followed by the validation of this neural network using more than 21 million\\nmolecules from the ZINC15 database. As a result, three small drug-like compounds that exhibited the\\nhigh-affinity binding to gp120 were identified. According to the data from molecular docking, machine\\nlearning, quantum chemical calculations, and molecular dynamics simulations, these compounds show\\nthe low values of binding free energy in the complexes with gp120 similar to those calculated usingthe same computational protocols for the HIV-1 entry inhibitors NBD-11021 and NBD-14010, highly\\npotent and broad anti-HIV-1 agents presenting a new generation of the viral CD4 antagonists. The\\nidentified CD4-mimetic candidates are suggested to present good scaffolds for the design of novelantiviral drugs inhibiting the early stages of HIV-1 infection.ARTICLE HISTORY\\nReceived 5 September 2020\\nAccepted 26 February 2021\\nKEYWORDS\\nHIV-1; gp120; HIV-1 entryinhibitors; deep learning;generative adversarial\\nautoencoder; virtual\\nscreening; moleculardocking; quantum chemicalcalculations; molecular\\ndynamics simulations;\\nbinding free energycalculations;anti-HIV-1 drugs\\nIntroduction\\nModern methods of computer-aided drug design significantly\\nexpand the capabilities of the pharmaceutical industry, vastlyreducing the time and costs required for developing noveltherapeutic agents (Katsila et al., 2016 ; Knight-Schrijver et al.,\\n2016 ; Leelananda & Lindert, 2016 ; Mallipeddi et al., 2014 ;\\nSliwoski et al., 2014 ; Wadood et al., 2013 ). Despite the effi-\\ncacy of computational methods in drug design is currently\\nuniversally recognized, the development of new mathemat-ical approaches and the availability of powerful and cheapcomputing resources promote their continuous improve-ment. Among these approaches, machine learning methodsand, in particular, deep learning techniques, which offersgreat potential for further progress in this research area,occupy an important place (Dobchev et al., 2014 ; Lipinski\\net al., 2019 ; Vamathevan et al., 2019 ). To date, computer-\\naided design of potential drugs using machine learningmethods is one of the most important and rapidly develop-ing areas of chem- and bioinformatics (Dobchev et al., 2014 ;\\nLipinski et al., 2019 ; Vamathevan et al., 2019 ). Unlike physical\\nmodels based on explicit physical equations, such asquantum chemistry or molecular dynamics (MD) modeling,\\nmachine learning approaches use pattern recognition algo-rithms to determine the mathematical relationships betweenempirical observations of small molecules and their extrapo-\\nlation to predict chemical, biological and physical properties\\nof new compounds (Dobchev et al., 2014 ; Lipinski et al.,\\n2019 ; Vamathevan et al., 2019 ). In addition, compared to\\nphysical models, machine learning methods are more effi-cient and can easily scale to big data sets. One of the bene-fits of using machine learning to design drugs is to helpresearchers understand and use the relationships between\\nchemical structures and their biological activity (Dobchev\\net al., 2014 ; Lipinski et al., 2019 ; Vamathevan et al., 2019 ).\\nModern machine learning methods can be used to modelthe quantitative structure-activity relationship (QSAR) orquantitative structure-property relations (QSPR) and developintelligent tools that can accurately predict the effect ofchemical modifications of a compound on its biological activ-\\nity, pharmacokinetic and toxicological characteristics\\n(Cherkasov et al., 2014 ; Verma et al., 2010 ). In addition, these\\nmethods can be applied for extraction of descriptors fromchemical structures automatically and directly (Duvenaud\\nCONTACT Alexander M. Andrianov alexande.andriano@yandex.ru Institute of Bioorganic Chemistry, National Academy of Sciences of Belarus, Minsk,\\nRepublic of Belarus\\n/C2232021 Informa UK Limited, trading as Taylor & Francis GroupJOURNAL OF BIOMOLECULAR STRUCTURE AND DYNAMICS\\n2022, VOL. 40, NO. 16, 7555 –7573\\nhttps://doi.org/10.1080/07391102.2021.1905559et al., 2015 ; Jaeger et al., 2018 ; Kadurin, Nikolenko et al.,\\n2017 ), drug repurposing (Agastheeswaramoorthy &\\nSevilimedu, 2020 ; Kinnings et al., 2011 ), protein structure pre-\\ndiction (Senior et al., 2020 ), virtual screening of drug candi-\\ndates and prediction of protein/ligand binding affinity (Chenet al., 2018 ; Ekins, 2016 ; Jimenez-Luna et al., 2020 ; Lavecchia,\\n2019 ; Li et al., 2021 ; Stokes et al., 2020 ; Xiong et al., 2020 ;\\nYang et al., 2019 ), and related aspects.\\nBesides these prediction applications, machine learning\\ncan also be used for generation of novel compounds with\\ndefined structural and functional properties (Goodfellowet al., 2014 ; Gupta et al., 2018 ; Olivecrona et al., 2017 ;\\nSanchez-Lengeling et al., 2017 ; Silver et al., 2016 ; Yu et al.,\\n2016 ; Zhang et al., 2020 ). Over the last 5 years, deep learning\\ngenerative models have emerged as novel methods for de\\nnovo drug design. Due to the tremendous progress of deep\\nlearning techniques, generative deep learning models with\\ndifferent architectures, data types, and learning methods thatintroduce an encouraging solution to drug discovery have\\nbeen currently developed (Ar /C19us-Pous, Blaschke et al., 2019 ;\\nAr/C19us-Pous, Johansson et al., 2019 ; Bian et al., 2019 ; Blaschke\\net al., 2020 ;G/C19omez-Bombarelli et al., 2018 ; Guimaraes et al.,\\n2017 ;M/C19endez-Lucio et al., 2020 ; Mercado et al., 2020 ;\\nPrykhodko et al., 2019 ; Segler et al., 2018 ; Zhavoronkov et al.,\\n2019 ). The recent applications of the generative deep learn-\\ning models clearly showed their ability to generate molecules\\nthat can be synthesized, are active in vitro , metabolically sta-\\nble, and exhibit in vivo activity in disease-relevant models. In\\nparticular, the Janus kinase 3 inhibitor (JAK3) presenting a\\nnew class of immunomodulatory agents was newly devel-\\noped using the conditional adversarial autoencoder\\n(Polykovskiy et al., 2018 ). Furthermore, in vivo active inhibi-\\ntors of discoidin domain receptors 1 and 2 (DDR1 and DDR2)\\nwere designed using Generative Tensorial ReinforcementLearning and the pharmacokinetic profile of DDR1 was vali-\\ndated by in vivo mouse experiments (Zhavoronkov et al.,\\n2019). In this regard, the development and application ofdeep learning generative methods for computer-aided\\ndesign of potential drugs is of great scientific and practical\\nimportance (Bian et al., 2019 ; Dobchev et al., 2014 ; Lipinski\\net al., 2019 ;M/C19endez-Lucio et al., 2020 ; Vamathevan et al.,\\n2019 ; Zhang et al., 2020 ; Zhavoronkov et al., 2019).\\nAn essential role in structure-based drug discovery\\nbelongs to molecular docking that is widely used to predict\\nthe ligand conformation and its position and orientation\\nwithin the binding site of target proteins, assess the bindingaffinity and investigate the interaction profile of drug candi-dates (Guedes et al., 2018 ; Meng et al., 2011 ). The latest\\ndevelopments of semiempirical quantum mechanical and\\ndensity functional theory methods as well as applications ofexplicit quantum mechanical calculations to structure-based\\ndrug design in the context of identification and optimization\\nof drug candidates show the growing importance of quan-tum chemistry in the study of protein-ligand interaction\\n(Cavasotto et al., 2018 ; Ryde & S €oderhjelm, 2016 ; Yilmazer &\\nKorth, 2016 ). Molecular dynamics is also a powerful approach\\nfor structure-based drug discovery (Hollingsworth & Dror,\\n2018 ). Unlike molecular docking, MD simulates moving eachatom separately in the field of the rest atoms and represents\\nthe flexibility of both the ligand and protein more effectivelythan other algorithms. Examples of successful applicationsclearly demonstrate the power of computational approachesto identify compounds with desired properties from large\\ndatabases and design novel small-molecule drug candidates\\n(Hollingsworth & Dror, 2018 ).\\nIn recent years, a large number of studies have appeared\\non the use of machine learning methods to predict potential\\nHIV-1 inhibitors and virus resistance to anti-HIV drugs\\n(reviewed in Dubey, 2018 ). However, all of these studies\\nfocus on viral enzymes, namely reverse transcriptase, prote-ase and integrase (Dubey, 2018 ; Qureshi et al., 2018 ).\\nCompounds that block these enzymes cannot prevent the\\nvirus from entering a target cell, which raises attention to\\nthe HIV-1 inhibitors that can interfere with the initial stagesof the HIV-1 infection cycle by blocking the viral adsorptionto CD4\\nþcells or/and the virus-cell membrane fusion\\n(Andrianov et al., 2017 ,2019 ; De Clercq, 2005 ; Este & Telenti,\\n2007 ; Li et al., 2017 ; Rusconi et al., 2007 ; Ryser & Fluckiger,\\n2005 ; Su et al., 2017 ).\\nAt the first stage of replication cycle, the HIV-1 envelope\\ngp120 protein binds to the primary cellular receptor CD4 and\\nthen to the chemokine co-receptors CCR5 and/or CXCR4\\n(Wilen et al., 2012 ). Sequential interactions of gp120 with CD4\\nand CCR5/CXCR4 trigger structural changes in the HIV-1 gp41transmembrane domain, resulting in the Env-mediated mem-brane fusion and viral entry into CD4 þcells (Wilen et al.,\\n2012 ). The HIV-1 entry into the host cell is therefore a highly\\npromising target for several types of antiviral agents, such asinhibitors of the gp120 /C0CD4 interaction, antagonists of the\\nco-receptors CCR5 and CXCR4, and inhibitors of the virus-cellmembrane fusion (Andrianov et al., 2017 ,2019 ).\\nAt present, many HIV-1 entry/fusion inhibitors have been\\ndeveloped (Andrianov et al., 2017 ,2019 ; De Clercq, 2005 ;E s t e\\n& Telenti, 2007 ;L ie ta l . , 2017 ; Rusconi et al., 2007 ;R y s e r&\\nFluckiger, 2005 ;S ue ta l . , 2017 ), but only two of these antiviral\\nagents are currently used in clinical practice for treatment of\\nthe HIV-1 infection. These are maravirok that binds to co-receptor CCR5 (MacArthur & Novak, 2008 ) and a peptide\\ninhibitor enfuvirtide blocking the gp41-mediated virus fusion(Matthews et al., 2004 ). However, like the anti-HIV-1 drugs\\nthat act inside the target cell, these compounds have a num-\\nber of limitations and severe side effects (Arts & Hazuda, 2012 ;\\nKumari & Singh, 2012 ), which makes it urgent to search for\\nnew potent HIV-1 entry inhibitors with acceptable ADMET(absorption, distribution, metabolism, excretion, and toxicity)\\nparameters (Antoine et al., 2017 ;K e r n s&D i , 2008 ).\\nBecause the HIV-1 gp120 binding to CD4 starts the process\\nof viral entry into susceptible cells, one of the most promisingareas in the design of novel antiviral drugs involves the search\\nfor small-molecule compounds able to break the key interac-\\ntions between gp120 and CD4. Despite numerous reports ofnew CD4-mimetic candidates, there are currently no licensedanti-HIV-1 drugs mimicking cellular receptor CD4 (reviewed inLi et al., 2017 ). In this connection, studies on the development\\nof selective and potent drug-like compounds blocking\\ngp120 /C0CD4 interface still cause great interest.7556 A. M. ANDRIANOV ET AL.The goal of this work was to develop and apply a genera-\\ntive adversarial autoencoder to identify potential HIV-1 entryinhibitors that target the hydrophobic Phe-43 cavity ofgp120 critically important for the virus attachment to theCDþT cells (Kwong et al., 1998 ).\\nIn doing so, the following studies were carried out: (i)\\narchitecture of the generative adversarial autoencoder wasconstructed; (ii) a virtual compound library of potential anti-HIV-1 agents for training the neural network was formed by\\nthe concept of click chemistry allowing one to generate a\\nlarge number of drug candidates by their assembly fromsmall modular units (Hein et al., 2008 ; Kolb et al., 2001 ;\\nMoses & Moorhouse, 2007 , p. 32; Thirumurugan et al., 2013 );\\n(iii) molecular docking of all compounds from this librarywith gp120 was made and calculations of the values of bind-\\ning free energy were performed; (iv) molecular fingerprints\\nfor chemical compounds from the training dataset were gen-erated; (v) training the neural network was implemented fol-lowed by validation of the learning outcomes and work ofthe autoencoder; (vi) methods of molecular docking were\\nused to select the most promising anti-HIV-1 drug candidate\\nfrom the ensemble of compounds, which were identified bythe fingerprint-based screening of the virtual compounddatabase when the autoencoder validation; (vii) the ligand-binding affinity of the identified drug candidate to gp120\\nwas predicted in terms of the scoring functions of molecular\\ndocking, machine learning, semiempirical quantum chemis-try, and molecular dynamics; (viii) the values of bindingenergy approximated by these scoring functions were com-pared to those predicted using the same computational pro-tocols for the HIV-1 inhibitors NBD-11021 and NBD-14010,\\nhighly potent and broad anti-HIV-1 agents presenting a new\\ngeneration of viral entry antagonists (Acharya et al., 2015 ;\\nCurreli et al., 2015 ,2017 , 2018, 2020).\\nMethods and materials\\nDevelopment of a generative adversarial autoencoder\\nAs the most commonly used machine learning models oper-\\nate with a fixed input size, the generative network beingdeveloped needed to convert a huge variety of molecular\\nstructures into the form of data appropriate for the autoen-\\ncoder under development. For this reason, the MACCS fin-gerprints ( http://www.dalkescientific.com/writings/NBN/finger\\nprints.html ;https://www.rdkit.org/ ) were chosen in this study\\nas the type of descriptors to train the model, given the suc-\\ncessful example of applying adversarial autoencoder for gen-\\nerating fingerprints encoding structures of potentialanticancer agents (Kadurin, Aliper et al., 2017 ). That is why\\nthis neural network was used in our study as a basic genera-tive model to solve the task.\\nFigure 1 sheds light on the architecture of the developed\\nadversarial autoencoder. The model consists of two neuralnetworks including an autoencoder and a discriminator thatwork during training in an adversarial mode ( Figure 1 ). This\\nallows one to configure the autoencoder parameters in the\\ntraining mode and provide obtaining a high-quality output\\ndata at the next stage of their generation. The task of thediscriminator is to distinguish real data from those generated\\nby the encoder. The developed autoencoder is a seven-layer\\nneural network with input and output layers, a latent layer,and four fully connected layers ( Figure 1 ).\\nFingerprints of chemical compounds are fed to the input\\nlayer, the data of which pass through two fully connected\\nlayers (encoder) and fall on the latent layer where a numer-\\nical estimate of the binding energy to the molecular target isadded to the result obtained. Next, the fingerprints pass\\nthrough two fully connected layers (decoder) and go to the\\noutput, which, like the input, is a fingerprint vector. Thelatent layer consists of three neurons two of which receive\\nvalues from the encoder, and the third one receives the\\nvalue of the binding free energy. In the autoencoder genera-tive mode, random numbers are entered to the two neurons\\nof the latent layer and a binding free energy value to the\\nthird neuron, and then pass through the decoder generatingfingerprints of molecules satisfying a pre-defined threshold\\nenergy value. For the generation of such molecules, it is\\nimportant that the data entering the latent layer after pass-ing through the encoder have a normal distribution. To\\nensure this condition, in the course of adversarial training of\\nthe encoder and discriminator, it was aimed that the encoderwas able to produce data with a normal distribution on thelatent layer, and that the discriminator to distinguish the\\nstandard normal distribution (generated data) from the distri-\\nbution obtained on the latent layer.\\nAs noted before, the developed adversarial autoencoder is\\nbased on the model of the neural network that wasdesigned to generate chemical compounds with anticancer\\nproperties (Kadurin, Aliper et al., 2017 ), but has the following\\nspecific features ( Figure 1 ):\\n/C15Instead of a neuron liable for percent growth inhibition\\nof cells (Kadurin, Aliper et al., 2017 ), a neuron responsible\\nfor the values of binding free energy is added to the\\nlatent layer. This neuron does not interact with theencoder and is fed only to the input of the decoder\\ntogether with the data obtained by the encoder. At the\\nlearning stage, the values of this neuron are assignedaccording to the estimation of value of binding free\\nenergy for the testing compounds. At the generative\\nstage, its values have the meaning of a binding freeenergy threshold. The latent layer contains three neurons;\\n/C15The encoder consists of two consecutive layers L1 and L2\\nwith 32 and 16 neurons, respectively. The decoder\\nincludes two layers L ’1 and L ’2 containing 16 and 32 neu-\\nrons, respectively;\\n/C15The discriminator consists of four layers including 2, 16, 3\\nand 1 neurons, respectively;\\n/C15The activation function leaky relu (Xu et al., 2015 ) is used\\non the intermediate layers of the autoencoder;\\n/C15Sigmoid activation functions are used on all layers of the\\ndiscriminator (Gibbs & MacKay, 2000 );\\n/C15To provide an additional level of protection against over-\\nfitting, a dropout layer is used between two fully con-nected layers of the encoder and decoder, allowing theJOURNAL OF BIOMOLECULAR STRUCTURE AND DYNAMICS 7557neural network to completely use its parameters (weights)\\nand selective random disconnection during the training.\\nA three-stage iterative procedure was used for training\\nthe developed adversarial autoencoder, namely: (i) trainingthe discriminator to distinguish a given normal distribution\\nfrom the encoded one obtained by the encoder on the\\nlatent layer; (ii) joint training of the encoder and decoder asan autoencoder; (iii) training the encoder to compress the\\ndata in such a way that they look like a normal distribution.\\nDiscriminatory models are characterized by a simpler\\nlearning process than generative neural networks. Therefore,\\nduring the first attempts to train the model, a situation\\noccurred when the function of the discriminator lossesdecreased, and the function of the encoder losses grew\\n(Figure 2 ).\\nIt should be noted that when training these two networks\\nin an adversarial mode, it is difficult to determine whether\\nthe encoder is really poorly trained or it can not ‘trick ’the\\ndiscriminator. Small discriminator losses combined with bigencoder losses do not necessarily indicate deficiencies in the\\ntraining of the encoder. Nevertheless, to improve the learn-\\ning process, the following measures have been taken:/C15Preliminary training of the discriminator before training\\nthe entire adversarial autoencoder. Preliminary training\\nwas carried out under the assumption that the discrimin-ator would distinguish random numbers obtained from\\nthe untrained encoder from those corresponding to the\\nnormal distribution used;\\n/C15When training the entire network, the discriminator was\\ntrained not every epoch, but once in 2 epochs, and the\\nautoencoder was trained every epoch;\\n/C15For the discriminator, a lower learning rate (0.001) was\\nset, while for the entire adversarial autoencoder it was\\nequal to 0.005;\\n/C15A small noise was added to the data generated from the\\nnormal distribution, which made the discriminator oper-\\nation more difficult.\\nThese changes concerned each epoch of training which\\npresented an iterative process including the following steps:\\nStep 1 /C0the encoder and decoder were trained jointly as\\nthe autoencoder (Creswell et al., 2018 );\\nStep 2 /C0the discriminator was trained to distinguish the\\ngiven normal distribution from the encoded representation\\nobtained by the encoder on the latent layer;\\nFigure 1. The architecture of a generative adversarial autoencoder for prediction of potential anti-HIV-1 agents targeting the viral envelope gp120 protein . The\\nencoder consists of two consecutive layers L1 and L2 with 32 and 16 neurons, respectively. The decoder includes two layers L ’1 and L ’2 containing 16 and 32 neu-\\nrons, respectively. The latent layer consists of 3 neurons.7558 A. M. ANDRIANOV ET AL.Step 3 /C0the encoder was trained to compress the data in\\nsuch a way that they represented a normal distribution.\\nFor the above steps, the following loss functions\\nwere used:\\nStep 1 /C0The reconstruction loss function was presented\\nby a mean squared error between original fingerprints and\\nvalues decoded by the decoder part of the model. This func-\\ntion penalized the network for deviations of restored finger-prints from the input ones and was calculated using the\\nformula\\nRl¼1\\nnXn\\ni¼1ðyi/C0y0\\niÞ2,\\nwhere n is the number of bits in the fingerprint, y iis the\\nvalue at the i-th position of the original fingerprint, y ’iis the\\nvalue at the i-th position of the fingerprint restored after\\ndecoding the values compressed by the encoder.\\nStep 2 /C0The discriminator loss function consisted of two\\nBinary Cross Entropy losses, where Binary Cross Entropy loss\\nwas (Dumoulin et al., 2017 ):\\nBCE label ,y ðÞ ¼/C0 ð label/C3logyðÞ/C01/C0label ðÞ /C3logð1/C0yÞÞ,\\nwhere label 2f0,1g–classes, and y is the discriminator ’s out-\\nput, predicted class probability.\\nThe first binary cross entropy loss related to discrimin-\\nator positive loss (the ability to correctly identify data gen-\\nerated from real normal distribution) and took as\\narguments ones and discriminator ’s prediction for values\\nsampled from real distribution. The second binary cross\\nentropy loss related to discriminator negative loss (the abil-\\nity to distinguish encoded data from real distribution). Ittook as arguments zeros and discriminator ’s output for\\nencoded values.\\nThe total equation used for calculation of the discrimin-\\nator loss was\\nDLF¼BCEð1,DðNDÞÞ þ BCEð0,DðEDÞÞ,\\nwhere D(ND) is the discriminator predicted value for the nor-\\nmal distribution, and D(ED) is the discriminator output for\\nthe encoded data (Janocha & Czarnecki, 2017 ).\\nStep 3 /C0The encoder loss function exposed the ability of\\nencoder to deceive discriminator and it was represented by\\na binary cross entropy loss between ones and encoded val-\\nues that were passed through discriminator. This functionwas calculated using the formulaELF¼BCEð1,DðEDÞÞ:\\nThe discriminator and autoencoder were trained jointly in\\ntwo stages: reconstruction stage (step 1) and regularization\\nstage (steps 2 and 3) which were performed in each subsetfrom the original data. At the reconstruction stage, theautoencoder updated the encoder and decoder to minimizethe recovery error of the input and output data. At the regu-larization stage, the discriminator network was first updated(step 2) to distinguish the true samples obtained by the nor-mal distribution generator from the compressed input dataon the latent layer calculated by the autoencoder, and then(step 3) the autoencoder updated its encoder to confuse thediscriminator network.\\nFigure 3 shows the graphs of the loss functions taking\\ninto account the changes made above. The discriminator lossfunction graph was further divided into two graphs present-ing the loss function of the correct classification of the nor-mal distribution and the loss function for the encoded data.As follows from Figure 3 , the discriminator was well trained\\nto classify data from a normal distribution, but its ability tocorrectly identify the data received by theencoder decreased.\\nThe following parameters were used to train the\\nautoencoder:\\n/C15The number of epochs for the main version of the model\\nused for generation was 400.\\n/C15The learning rate of the entire autoencoder in step 1\\nwas 0.005.\\n/C15The learning rate of the discriminator in step 2 was 0.001.\\n/C15The encoder learning rate at step 3 was 0.005.\\n/C15The Batch size parameter was set to 128.\\n/C15Optimizer used the Adam method (Kingma & Ba, 2014 ).\\nWhen training the neural network, it was important to\\nmake sure that in the generation mode the model is able tomake different results for various input data and producemany molecular fingerprints. At the same time, this modelwas not supposed to generate fingerprints from a smallnumber of possible options, since there was a probabilitythat such a situation reflected some minimum of the initialloss functions. In order to reduce the probability of such ascenario, the tSNE algorithm (Van der Maaten & Hinton,2008 ) was used to subsample the generated molecular\\nFigure 2. The discriminator and encoder loss functions with faster discriminator training. The graphs show the informative part of the functions.JOURNAL OF BIOMOLECULAR STRUCTURE AND DYNAMICS 7559fingerprints ( Figure 4 ). Each generated fingerprint represents\\nmolecular descriptor as a two- or three-dimensional point in\\nsuch a way that similar molecular descriptors are displayed\\nby closely spaced points, and dissimilar descriptors are dis-\\nplayed with a high probability by points that are far from\\neach other.\\nIt should be noted that, due to the adversarial process,\\nthe autoencoder was learned to compress the input data in\\nsuch a way that it enabled similar compounds to have\\nresembling latent vector representations, providing thereby\\ngeneration of meaningful samples in the generation mode.\\nGeneration of the training dataset\\nThe training dataset was generated by the AutoClickChem\\nsoftware package (Durrant & McCammon, 2012 ). This pro-\\ngram implements the click chemistry methodology in silicoto design a large set of drug candidates that include the\\nmoieties providing the ligand with biological activity towards\\na target protein. In the first step, two compound libraries\\ncontaining small molecules with the functional groupsinvolved in the click-reaction of azide-alkyne cycloaddition\\n(Kolb et al., 2001 ) were created by virtual screening of the\\nDrug-Like subset of the ZINC15 database, a virtual collectionof 1.5 billion molecules designed for in silico screening\\n(Sterling & Irwin, 2015 ). For this purpose, the DataWarrior\\nprogram (Sander et al., 2015 ) was used. Aromatic compounds\\nwith molecular mass <250 Da containing azide or alkyne\\ngroups were selected and placed into library 1, and all low-\\nmolecular compounds (molecular mass <250 Da) with the\\nabove functional groups were collected in library 2. The\\nselection of aromatic compounds as a scaffold for the design\\nof potential CD4-mimetic candidates is explained by the factthat their aromatic moieties can mimic the key interactions\\nof the Phe-43 residue of CD4 with the hydrophobic Phe-43\\nFigure 3. The discriminator loss function (DLF) consisting of the DLF for normal distribution and the DLF for encoded data, and the encoder loss function. The\\ngraphs show the informative part of the functions.\\nFigure 4. The results of the tSNE algorithm action for three (a) and two (b) components with a subsample from the generated molecular fingerprints.7560 A. M. ANDRIANOV ET AL.cavity of gp120. As follows from the crystal gp120 /C0CD4\\ncomplex (Kwong et al., 1998 ), the aromatic ring of Phe-43 CD4\\nis buried into the Phe-43 cavity of gp120, leading to blocking\\nthe gp120 residues critical for the viral adsorption to CD4 þT\\ncells. Furthermore, novel potent viral entry antagonists, suchas NBD-11021, NBD-14010 and NBD-14189, show the similarinteraction modes to bind this hydrophobic pocket of gp120\\n(Acharya et al., 2015 ; Curreli et al., 2015 ;2017 ; 2018; 2020).\\nIn sum, libraries 1 and 2 contained 1388 and 3769 com-\\npounds, respectively. These compounds were then used as\\nreactants to mimic the click-reaction of azide-alkyne cyclo-\\naddition by AutoClickChem (Durrant & McCammon, 2012 ),\\nleading to 1 655 301 hybrid molecules. 120 000 compounds\\nout of the designed molecules that fully satisfied the\\nLipinski ’s‘rule of five ’(Lipinski et al., 2001 ) were included\\ninto the training dataset.\\nMolecular docking of these drug-like molecules with\\ngp120 was performed by the QuickVina 2 program package\\n(Alhossary et al., 2015 ) in the approximation of rigid receptor\\nand flexible ligands. The gp120 structure was isolated from\\nthe crystal gp120 /C0CD4/C0Ab17b complex (Kwong et al.,\\n1998 ). The gp120 and ligand structures were prepared by\\nadding hydrogen atoms with the OpenBabel software\\n(http://openbabel.org/wiki/Main_Page ) followed by their opti-\\nmization in the UFF force field (Rappe et al., 1992 ). The grid\\nbox included the hydrophobic Phe-43 cavity of gp120 andwas the site of the X-ray structure (Kwong et al., 1998 ) with\\nthe following parameters: DX¼10 Å, DY¼10 Å, DZ¼10 Å\\ncentered at X¼29 Å, Y ¼/C0 10 Å, Z ¼/C0 26 Å. The value of\\n‘exhaustiveness ’parameter defining number of individual\\nsampling ‘runs ’was set to 50 (Alhossary et al., 2015 ).\\nThe values of scoring function calculated for the analyzed\\nmolecules bound to gp120 varied in the range from\\n/C02.5 kcal/mol to /C08.5 kcal/mol, which indicates that these\\ncompounds show different binding affinities to the targetprotein and, hence, form a suitable dataset to train theautoencoder.\\nMolecular fingerprints of the compounds from the train-\\ning dataset /C0MACCS (Molecular ACCess System) keys ( http://\\nwww.dalkescientific.com/writings/NBN/fingerprints.html )/C0\\nwere calculated using the open-source chemoinformatics\\nsoftware package RDKit ( https://www.rdkit.org/ ). The\\nMACCS166KeyBits value of Type option of the MACCS keysrepresenting molecular fingerprint by a bit-vector string of\\n166 bits containing zeros and ones was used ( http://www.dal-\\nkescientific.com/writings/NBN/fingerprints.html ;https://www.\\nrdkit.org/ ).\\nApplication of the developed autoencoder for anti-HIV-1\\ndrug prediction\\nApplying the developed autoencoder for anti-HIV-1 drug pre-\\ndiction proposes an operation of this neural network in com-\\nbination with similarity-based virtual screening andtraditional methods of molecular modeling. A general\\nscheme of the algorithm for anti-HIV-1 drug prediction is\\nshown in Figure 5 .The decoder generates a set of the MACCS structural keys\\nat a given threshold value of the binding energy to screen\\nthe fingerprint library of the Drug-Like compounds fromZINC15, which enables one to significantly reduce the\\namount of data considering in the process of virtual screen-\\ning of the library of molecular fingerprints. Compoundsexhibiting fingerprints similar to those generated by the\\nautoencoder are then docked with the HIV-1 envelope\\ngp120 protein using molecular docking. Next, the docked lig-and/gp120 complexes are optimized by the semiempirical\\nquantum chemical method PM7 to improve the accuracy of\\nprediction of interaction modes obtained based on classicalforce field. At the final stage, molecular dynamics simulations\\nof the identified compounds bound to gp120 are performed\\nfollowed by the post-modeling analysis of the MD trajecto-\\nries, including prediction of the possible interaction modes\\nand binding affinity profile for the dynamic ligand/gp120 structures.\\nValidation of the adversarial autoencoder\\nTo validate the robustness of the developed autoencoder,the library of the MACCS keys ( http://www.dalkescientific.\\ncom/writings/NBN/fingerprints.html ) for 21 325 567 com-\\npounds from the Drug Like subset of the ZINC15 database\\n(Sterling & Irwin, 2015 ) was constructed using the RDKit soft-\\nware package ( https://www.rdkit.org/ ). Then, molecular fin-\\ngerprints of the five ligands were generated by the\\nautoencoder at the threshold energy values of /C05.0 kcal/mol\\nand /C08.0 kcal/mol. As a result of virtual screening of the\\nMACCS keys library, compounds with fingerprints similar to\\nthose generated by the neural network were found. As ameasure of fingerprint similarity, the Hamming distance\\n(Hamming, 1950 ) that is defined in coding theory as the\\nnumber of pairs of mismatched components in the com-pared vectors and the Tanimoto coefficient (Sharma & Lal,\\n2011 ; Tanimoto, 1957 ) were used.\\nThe Tanimoto coefficient was calculated with the formula\\n(Sharma & Lal, 2011 ; Tanimoto, 1957 )\\nTa,bðÞ ¼\\nNc\\nNaþNb/C0Nc,\\nwhere T is the Tanimoto coefficient taking values from 0 to\\n1; N ais the number of elements in the first vector; N bis the\\nnumber of elements in the second vector; N cis the number\\nof identical elements in two vectors. When screening the\\nMACCS keys library for a given fingerprint, compounds from\\nthe ZINC15 database for which the Tanimoto coefficient sat-\\nisfied the criterion T>0.85 were selected (Sharma & Lal,\\n2011 ; Tanimoto, 1957 ).\\nIn line with the general scheme given in Figure 5 , the\\ncompounds identified in the ZINC15 database (Sterling &\\nIrwin, 2015 ) were docked with the crystal gp120 structure\\n(Kwong et al., 1998 ) and the values of binding free energy\\nwere calculated using QuickVina 2 (Alhossary et al., 2015 ).\\nThe molecular docking was performed by the computational\\nprotocol identical to that applied for generation of the train-ing dataset (see above). Besides, the docking models of theJOURNAL OF BIOMOLECULAR STRUCTURE AND DYNAMICS 7561top-ranking compounds in terms of the QuickVina scoring\\nfunction were optimized by the semiempirical quantum\\nchemical method PM7 (Stewart, 2013 ) to improve the ligand\\npositioning accuracy (Sulimov et al., 2017 ). The quantum\\nchemical optimization of the docked structures was carriedout in the MOPAC2016 package ( http://OpenMOPAC.net ).\\nBefore the calculations, hydrogen atoms were added to the\\ncomplexes of interest followed by their energy refinement inthe UFF force field (Rappe et al., 1992 ). To do this, the Open\\nBabel program ( http://openbabel.org/wiki/Main_Page ) was\\nemployed. The calculations were carried out by the COSMOsolvation model (COnductor-like Screening MOdel) (Klamt,\\n2005 ; Klamt & Sch €u€urmann, 1993 ) using an implicit solvent\\nwith water ’s dielectric constant of 78.4 ( http://OpenMOPAC.\\nnet). The Localized Molecular Orbitals method (Høyvik et al.,\\n2012 ) available in MOPAC in the form of the linear scaling\\nSCF MOZYME algorithm (Stewart, 2013 ;http://OpenMOPAC.\\nnet) was used to speed up the calculations. The value of\\nRMS gradient was set to 20 kcal/mol//C19Å( http://\\nOpenMOPAC.net ).\\nThe classical dynamics of the ligand/gp120 complexes in\\nwater was made with the implementation of Amber18 using\\nthe Amber ff14SB force field (Case et al., 2020 ). The\\nAntechamber module was employed to set the Gasteigeratomic partial charges (Case et al., 2020 ). To prepare the\\nforce field parameters, the general Amber GAFF force field(Wang et al., 2004 ) was used. Hydrogen atoms were added\\nto gp120 by the tleap program of the AmberTools18 pack-\\nage (Case et al., 2020 ). Initially, the ligand/gp120 complexes\\nwere each placed in a cubical box with periodic boundaryconditions. In addition to the ligand/gp120 complex, the boxfor the MD simulations included TIP3P water (Jorgensenet al., 1983 ) as an explicit solvent, Na\\nþand Cl/C0ions provid-\\ning overall salt concentration of 0.10 M. After setting up the\\nsystem, an energy minimization was performed using 500steps of the steepest descent algorithm followed by 500\\nsteps of the conjugate-gradient method. The backbone\\natoms of the complex assembly were then fixed by an add-itional harmonic potential with the force constant of 2.0 kcal/mol and the system was subject to the equilibration phase.The system equilibration was carried out in three consecutive\\nstages: 1) the system was gradually heated from 0 K to 310 K\\nfor 1 ns in NVT ensemble using a Langevin thermostat with acollision frequency of 2.0 ps\\n/C01(Case et al., 2020 ); 2) pressure\\nequilibration was made for 1 ns at 1.0 bar in NPT ensembleusing Berendsen barostat with a 2.0 ps characteristic time(Case et al., 2020 ); 3) the constraints on the complex assem-\\nbly were removed and the system was equilibrated again at\\n310 K over 0.5 ns under constant volume conditions. Afterequilibration was achieved, the MD simulations were carriedout for 50 ns in NPT ensemble at temperature T¼310 K and\\np¼1 bar. Bonds involving hydrogen atoms were constrained\\nusing SHAKE algorithm (Ryckaert et al., 1977 ) to achieve the\\nintegration time-step of 2 ps. Long-range electrostatic inter-actions were calculated using Particle Mesh Ewald (PME)algorithm (Essmann et al., 1995 ). Coulomb interactions and\\nvan der Waals interactions were truncated at 8 Å.\\nThe binding modes of the predicted compounds to\\ngp120, namely hydrogen bonds, salt bridges, van der Waalscontacts, and p-pinteractions were identified by the BINANA\\nprogram (Durrant & McCammon, 2011b ). The ligand poses in\\nthe docking ligand/gp120 models were visualized with theprogram UCSF Chimera (Pettersen et al., 2004 ). To visualize\\nvan der Waals contacts, the program LigPlot (McDonald &\\nThornton, 1994 ) was employed. The values of K\\ndfor the\\ndocking ligand/gp120 structures were calculated using aneural-network-based scoring function NNScore 2.0 (Durrant& McCammon, 2011a ). The values of binding free energy\\nwere estimated from those of K\\ndby the formula\\nDG¼R/C2T/C2ln(K d) (where DG is the binding free energy, R is\\nFigure 5. An operation scheme of the developed autoencoder for anti-HIV-1 drug prediction.7562 A. M. ANDRIANOV ET AL.the universal gas constant, T is the absolute temperature\\nequal to 310 K) (Sharma & First, 2009 ).\\nFor the PM7-based complexes, the ligand-binding affinity\\nwas estimated in terms of the values of binding enthalpy DH\\ncalculated as the differences between the heats of formationof the ligand/gp120 complexes and heats of formation of\\nthe ligand and gp120 in the unbound states (Stewart, 2013 ;\\nhttp://OpenMOPAC.net ). Quantum chemical optimization of\\nthe ligand and gp120 structures in the unbound states were\\nperformed using the computational protocol described\\nabove for the docking ligand/gp120 models.\\nIn the case of dynamic ligand/gp120 models, the values\\nof binding enthalpy were calculated with Amber18 (Case\\net al., 2020 ) using the MM/GBSA method (Genheden & Ryde,\\n2015 ; Sun et al., 2014 ; Xu et al., 2013 ). The calculations were\\nmade for 200 snapshots extracted from the final 40 ns of the\\nMD trajectories, by keeping the snapshots every 0.2 ns. The\\npolar solvation energies were computed in continuum solv-ent using Poisson-Boltzmann continuum-solvation model\\nwith ionic strength of 0.10. The non-polar terms were esti-\\nmated using solvent accessible surface areas (Case et al.,2020 ). The values of the entropic term of the binding free\\nenergy were calculated using the Nmode module of\\nAmberTools18 (Case et al., 2020 ). Analysis of the MD trajecto-\\nries was performed for the last 40 ns time domain by the\\nCPPTRAJ module of AmberTools18 (Case et al., 2020 ).\\nThe HIV-1 entry inhibitors NBD-11021 and NBD-14010,\\npresenting a new generation of viral entry antagonists thatblock gp120 /C0CD4 interaction with pan-neutralization of\\ndiverse subtypes of clinical isolates (Acharya et al., 2015 ;\\nCurreli et al., 2015 ,2017 ,2018 , 2020), were also involved in\\nthe calculations as a positive control.\\nImportantly, the training and testing of the developed\\nautoencoder were performed using the datasets containing\\ndifferent MACCS keys. This follows from the comparative\\nanalysis of fingerprints from the corresponding libraries indi-cating that only 331 fingerprints out of 120 0000 MACCS\\nkeys of the training library were identical to those of com-\\npounds from the Drug Like subset of the ZINC15 database.\\nThe generator part of the adversarial autoencoder for cre-\\nating molecular fingerprints with a preset value of bindingenergy is available at https://github.com/Nikishul/AAE-\\nsupplementary .\\nResults and discussion\\nThe results of the autoencoder validation are shown in\\nTables 1 and 2. In these Tables, the MACCS keys are repre-\\nsented by a string of 166 bits in which each bit position rep-resents the presence ( ¼1) or absence ( ¼0) of a pre-defined\\nstructural feature ( http://www.dalkescientific.com/writings/\\nNBN/fingerprints.html ). At the same time, for each finger-\\nprint, the codes of the top compounds according to the\\nHamming and Tanimoto criteria are given.\\nAnalysis of the obtained data shows ( Tables 1 and 2) that\\nthe combined use of the developed neural network with vir-\\ntual screening of the library of molecular fingerprints allowsone to identify the ligands with the lower values of bindingfree energy compared to a given threshold value.\\nFurthermore, the compounds with codes ZINC000026430653(Table 1 ), ZINC000191389930 and ZINC000293423658\\n(Table 2 ) in the ZINC15 database ( Figure 6 ) exhibit the values\\nof binding energy comparable with that of /C09.5 ± 0.1 kcal/\\nmol measured for the CD4 /C0gp120 complex by isothermal\\ntitration calorimetry (Myszka et al., 2000 ). These values are\\nalso very close to those predicted by QuickVina 2 for NBD-\\n11021 and NBD-14010 ( Table 3 ). The data on the high-affinity\\nbinding of ZINC000026430653, ZINC000191389930 andZINC000293423658 to gp120 are also supported by the val-\\nues of binding energy calculated for the PM7-based ligand/\\ngp120 models as the differences between the free energiesof the ligand/gp120 complexes and free energies of the lig-and and gp120 in the unbound states (Stewart, 2013 ;http://\\nOpenMOPAC.net )(Table 3 ). Furthermore, the above conclu-\\nsions are in agreement with the values of binding freeenergy of the analyzed molecules to gp120 calculated fromthe predicted values of K\\nd(Sharma & First, 2009 )(Table 3 ).\\nTaking together, the data of Table 3 suggest that the identi-\\nfied compounds may expose strong attachment to thehydrophobic Phe-43 cavity of the HIV-1 CD4-binding site ofgp120, in line with the low values of binding energy compar-\\nable with those predicted for NBD-11021 and NBD-14010.\\nExamination of the PM7-based ligand/gp120 complexes\\nreveals multiple van der Waals contacts of the identifiedcompounds with the gp120 Phe-43 cavity that plays a key\\nrole in the HIV-1 binding to cellular receptor CD4 ( Figures 7\\nand 8). By analogy with CD4 (Kwong et al., 1998 ), one of the\\naromatic systems of these compounds is buried into this\\nhydrophobic pocket and mimics the pivotal interactions of\\nthe benzene ring of Phe-43\\nCD4with the functionally import-\\nant residues of gp120. These gp120 residues are Val-255,Thr-257, Glu-370, Ile-371, Asp-368, Ser-375, Phe-382, Trp-427\\nand Met-475 ( Figures 7 and 8). From the crystal gp120 /C0CD4\\nstructure, all these amino acids are involved in the directinter-atomic contacts with CD4, and the highly conservedAsp-368, Glu-370 and Trp-427 are the dominant contributors\\nto the gp120 /C0CD4 interaction (Kwong et al., 1998 ).\\nAlong with van der Waals contacts, analysis of the PM7-\\nbased ligand/gp120 structures indicates a high probability ofa special type of weak hydrogen bond C –H/C1/C1/C1O (Desiraju &\\nSteiner, 1999 ; Weiss et al., 2001 ) associated with the CH\\ngroups of the aromatic rings of ZINC000026430653,ZINC000191389930, ZINC000293423658 and the carboxyl\\ngroup of Asp-368 that is located within the vestibule of the\\nPhe-43 cavity of gp120 ( Figure 7 ). These are evidenced by\\nthe values of the distance dbetween the O and C atoms\\nand the angle abetween the C, H and O atoms ( Figure 7 )\\nwhich support the assumption that these atoms may be\\ninvolved in a dipole-dipole C –H/C1/C1/C1O interaction (Steiner,\\n2003 ). These findings are of considerable interest because\\nhydrogen bonding with Asp-368 of gp120 is highly import-\\nant for potential CD4 mimetics to show the properties char-\\nacteristic for the small-molecule CD4 antagonists (Acharyaet al., 2015 ; Curreli et al., 2015 ,2017 , 2018, 2020).\\nThus, the data of molecular docking in conjunction with\\nquantum chemical calculations indicate that the identifiedJOURNAL OF BIOMOLECULAR STRUCTURE AND DYNAMICS 7563compounds may expose the high binding affinity to the\\nhydrophobic Phe-43 cavity of gp120 by the binding modes\\nmimicking the key interactions of CD4 with this HIV-1 enve-lope glycoprotein. However, it is well-known that estimatingthe binding affinity profiles using molecular docking studies\\ninvolves a number of approximations, resulting in the lack of\\nconfidence on the ability of scoring functions to give accur-ate values of binding free energy (Guedes et al., 2018 ;L i\\net al., 2021 ; Meng et al., 2011 ; Salmaso & Moro, 2018 ; Xiong\\net al., 2020 ). It is well known that conventional scoring func-\\ntions assume a predetermined theory-inspired functional\\nform for the relationship between the variables that charac-\\nterize the complex and its predicted binding affinity (Ainet al., 2015 ; Ballester & Mitchell, 2010 ; Brylinski, 2013 ; Guedes\\net al., 2018 ; Li et al., 2021 ; Xiong et al., 2020 ). Machine-learning scoring functions, which were developed as a prom-\\nising alternative to overcome the limitations of classical scor-\\ning functions, have been shown to outperform a wide rangeof classical scoring functions at both binding affinity predic-tion and virtual screening (Guedes et al., 2018 ; Li et al., 2021 ;\\nRagoza et al., 2017 ; Shen et al., 2021 ;W/C19ojcikowski et al.,\\n2017 ; Xiong et al., 2020 ). At the same time, it was also shown\\n(Shen et al., 2021 ) that the machine learning-based scoring\\nfunctions developed directly from classical scoring functions\\ndo not outperform those intrinsic ones, such as NNScore andRFScore. In this connection, it is important to note that the\\nvalues of binding free energy calculated for the docking lig-\\nand/gp120 models using the QuickVina scoring function arein good agreement with those of K\\nd(Table 3 ) predicted by\\nNNScore 2.0, a machine-learning scoring function which wasTable 1. Results of the autoencoder validation at a pre-defined threshold energy value of –5.0 kcal/mol.\\nMACCS keys of the ligands generated using the neural network Compound code in ZINC15Hamming distance (R);\\nTanimoto coefficient (T) Binding free energy, kcal/mol\\n000000000000000000000000000000\\n000000000000000000000100000000010010000100001010110010000000000101000100000000001000000001\\n110110000100101010000101100101\\n1010110101111110ZINC000026430653 R¼3;T¼0.96 /C08.8\\nZINC000037104033 R¼3T¼0.96 /C07.1\\nZINC000002786698 R¼4;T¼0.95 /C07.0\\nZINC000055836809 R¼3;T¼0.96 /C06.9\\nZINC000055843838 R¼4;T¼0.95 /C06.5\\n000000000000000000000000000000\\n000000000000000000000100000000\\n010010000100001010110010000000\\n0001111001000000000010000000011101100001001010100001011001010010111101111110ZINC000026430653 R¼5;T¼0.94 /C08.8\\nZINC000037104033 R¼6;T¼0.93 /C07.1\\nZINC000002786698 R¼6;T¼0.93 /C07.0\\nZINC000163393594 R¼4;T¼0.95 /C06.0\\nZINC000128895014 R¼4;T\\n¼0.95 /C05.9\\n000000000000000000000000000000\\n000000000000000000000100000000010010000100001010110010000000000111100100000000001000000001\\n110110010100000011000101100101\\n0010111101111110ZINC000035245594 R¼6;T¼0.93 /C06.6\\nZINC000163489237 R¼7;T¼0.92 /C06.6\\nZINC000052221501 R¼7;T¼0.92 /C06.6\\nZINC000600676089 R¼6;T¼0.93 /C06.4\\nZINC000004006242 R¼7;T¼0.92 /C05.9\\n000000000000000000000000000000\\n000000000000000000000100000000\\n010010000100001010110010000000\\n0001010001000000000010000000011101100001001000100001011001011010110101111110ZINC000026430653 R¼4;T¼0.95 /C08.8\\nZINC000002786698 R¼5;T¼0.94 /C07.0\\nZINC000055836809 R¼4;T¼0.95 /C06.8\\nZINC000037104033 R¼4;T¼0.95 /C06.7\\nZINC000685198234 R¼4;T¼0.95 /C0\\n6.0\\n000000000000000000000000000000\\n000000000000000000000100000000010010000100001010110010000000000111100100000000001000000001\\n110110010100101011000101100101\\n0010111111111110ZINC000182934853 R¼7;T¼0.92 /C07.4\\nZINC000771860139 R¼5;T¼0.94 /C06.6\\nZINC000012991344 R¼7;T¼0.92 /C06.0\\nZINC000128895014 R¼7;T¼0.92 /C06.0\\nZINC000163393499 R¼7;T¼0.92 /C06.0\\nTable 2. Results of the autoencoder validation at a pre-defined threshold energy value of –8.0 kcal/mol1,2.\\nMACCS keys of the ligands generated using\\nthe neural network Compound code in ZINC15Hamming distance (R);\\nTanimoto coefficient (T) Binding free energy, kcal/mol\\n0000000000000000000000000000000\\n000000000000000000001000000000100100001000010101100100000000001111001000000000010000000011101\\n1000010010001000010110010100101\\n11101111110ZINC000191389930 R¼6;\\nT¼0,93/C08,0\\n0000000000000000000000000000000\\n0000000000000000000010000000001\\n0010000100001010110010000000000\\n1111001000000000010000000011101100001001000110001011001110010111111111110ZINC000293423658 R¼7;\\nT¼0,92/C07,7\\n1The data for two of five MACCS keys selected for the autoencoder validation are given. For the other three structural keys, compounds with the values of\\nQuickVina 2 scoring function close or lower than the threshold energy value of /C08.0 kcal/mol were not found in ZINC15.\\n2For the second molecular fingerprint, compound ZINC000293423658 exhibiting the lowest value of binding energy among the identified molecules and c om-\\nparable with that of /C08.0 kcal/mol was selected for the further discussion.7564 A. M. ANDRIANOV ET AL.developed as an alternative to the classical scoring functions\\nof the commonly used docking programs AutoDock,\\nAutoDock Vina and several others (Durrant &\\nMcCammon, 2011a ).\\nAlthough optimization of docking ligand/protein models\\nby the semiempirical quantum chemical method PM7 signifi-\\ncantly improves the ligands positioning accuracy in the pro-tein active site (Sulimov et al., 2017 ), the use of this method\\nallows one to predict the binding free energies only in the\\napproximation of their enthalpic terms (Stewart, 2013 ;http://\\nOpenMOPAC.net ). The thermodynamic parameters based on\\nthe MD simulations are much more reliable than those pre-dicted from the analysis of static docking models and semi-\\nempirical quantum chemical studies. Unlike these\\ncomputational approaches, MD simulates moving each atom\\nseparately in the field of the rest atoms and represents the\\nflexibility of both the ligand and protein more effectively\\nthan other algorithms (Genheden & Ryde, 2015 ; Meng et al.,2011 ; Sakano et al., 2016 ; Salmaso & Moro, 2018 ; Sun et al.,\\n2014 ; Xu et al., 2013 ).\\nNevertheless, the MD simulations confirm the data on the\\nhigh binding affinity of the identified compounds to gp120\\nobtained from the post-modeling analysis of the static lig-\\nand/gp120 structures. The docked ligand/gp120 complexesreveal relative structural stability on the MD trajectories, asevidenced by the averages of binding free energy and theirstandard deviations ( Table 4 ). The low mean values of bind-\\ning free energy calculated for the dynamic models of theidentified compounds bound to gp120 ( Table 4 ) support the\\nabove assumption according to which all predicted CD4-mimetic candidates exhibit strong attachment to the CD4-binding site of gp120. Given the MM/GBSA standard error of1/C03 kcal/mol (Genheden & Ryde, 2015 ), the data obtained\\nsuggest that the averages of binding free energy for com-\\npounds I and III are comparable with the values predicted\\nfor NBD-11021 and NBD-14010 ( Table 4 ), and the corre-\\nsponding average for compound II is close to the experimen-tal value of /C09.5 ± 0.1 kcal/mol obtained for the complex of\\ngp120 with CD4 (Myszka et al., 2000 ). Finally, the values of\\nbinding enthalpy calculated for the dynamic ligand/gp120complexes ( Table 4 ) are similar to those estimated for their\\nstatic models by the semiempirical quantum chemicalmethod PM7 ( Table 3 ).\\nThe data on the time dependences of the root-mean\\nsquare deviations (RMSD) of the atomic positions betweenall of the MD ligand/gp120 structures and their starting mod-els also indicate the relative conformational stability of theCD4-mimetic candidates in the complexes with gp120. Asfollows from Figure 9 , these complexes do not subject to sig-\\nnificant structural changes within the MD simulations, as\\nFigure 6. Chemical structures of ZINC000026430653 (a), ZINC000191389930 (b) and ZINC000293423658 (c). For these molecules, the four Lipinski ’s parameter-\\ns/C0molecular mass, lipophilicity (LogP), number of H-bond donors and acceptors (Lipinski et al., 2001 )/C0are respectively: (a) 318.344 Da, 1.673, 0 and 8; (b)\\n252.281 Da, 1.487, 0 and 6 and (c) 312.333 Da, 1.544, 0 and 8.\\nTable 3. Values of binding energy ( DG) and K dcalculated for the static\\nligand/gp120 models.\\nCompoundDGDOCK1,\\nkcal/molKd2\\n(mM)DGKd3,\\nkcal/molDGPM74,\\nkcal/mol\\nZINC000026430653 /C08.8 0.788 /C08.65 /C039.17\\nZINC000191389930 /C08.0 1.061 /C08.46 /C040.78\\nZINC000293423658 /C07.7 9.591 /C07.11 /C038,84\\nNBD11021 /C08.4 0.948 /C08.53 /C029.58\\nNBD14010 /C08.9 0,039 /C010.49 /C042.77\\n1TheDG values according to the QuickVina 2 scoring function.\\n2The values of K dcalculated using a neural-network-based scoring function\\nNNScore 2.0.\\n3TheDG values estimated from those of K d.4TheDG values predicted based on the PM7-based ligand/gp120 complexes.JOURNAL OF BIOMOLECULAR STRUCTURE AND DYNAMICS 7565evidenced by the mean values of the RMSD calculated for\\nthe dynamic models of the predicted compounds in thecomplexes with gp120. These mean values and correspond-\\ning standard deviations, which are 2.2 ± 0.3 Å (compound I),\\n2.8 ± 0.6 Å (compound II), and 3.1 ± 0.9 Å (compound III), arecomparable with the averages of 2.9 ± 0.6 Å and 2.3 ± 0.4 Å\\ncalculated for NBD-11021 and NBD-14010, respectively\\n(Figure 9 ). Comparison of the MD structures between every\\nsubsequent dynamic model with the previous one results in\\nthe averages of RMSD which are also evidence of the relative\\nstability of the ligand/gp120 complexes during the MD simu-lations. For the identified compounds, these averages equalto 0.88 ± 0.08 Å (compound I), 0.89 ± 0.08 Å (compound II)\\nand 0.90 ± 0.07 Å (compound III) are very close to those cal-\\nculated for the control molecules NBD-11021 (0.91 ± 0.08 Å)and NBD-14010 (0.86 ± 0.07 Å).\\nDecomposition of the enthalpy components of binding\\nfree energy into the contributions of individual gp120 aminoacids exposes the residues dominating the ligand/gp120\\ninteractions in all of the cases of interest ( Table 5 ). AnalysisofTable 5 shows that these gp120 residues are Val-255, Thr-\\n257, Glu-370, Ile-371, Trp-427, and Met-475, which is consist-ent with the interaction modes observed in the static ligand/\\ngp120 models ( Figure 7 ). It is important to note that all\\nthese residues play a key role in the HIV-1 binding to CD4(Kwong et al., 1998 ). The data obtained suggest that there\\nare a number of the basic anchoring residues of gp120 that\\nfacilitate the analyzed compounds to effectively inhibit thepivotal interactions of the HIV-1 CD4-binding site with cellu-\\nlar receptor CD4. This supposition is supported by the data\\non the values of root-mean-square fluctuations (RMSF) of theindividual residues of gp120 indicating the flexibility of eachamino acid during the MD simulations ( Table 6 ). Analysis of\\nTable 6 shows that, according to the values of RMSF, the key\\nanchoring residues Val-255, Thr-257, Glu-370, Ile-371, Trp-427, and Met-475 of gp120 are positionally restrained on the\\nMD trajectories, in agreement with the data on their contri-\\nbutions into the binding enthalpy ( Table 5 ).\\nSo, the data on the ligand/gp120 binding affinities\\nobtained using four different scoring functions are consistent\\nFigure 7. The PM7-based complexes of ZINC000026430653 (a), ZINC000191389930 (b) and ZINC000293423658 (c) with gp120. The Phe-43 cavity of gp120 and the\\nresidues located within the vestibule of this hydrophobic pocket are shown. The residues of gp120 forming van der Waals contacts with the CD4-mimetic candi-\\ndates are indicated. C –H/C1/C1/C1O interactions of the CH groups of the aromatic rings of ZINC000026430653, ZINC000191389930 and ZINC000293423658 with the carb-\\noxyl group of Asp-368 gp120 are marked by dotted yellow lines.7566 A. M. ANDRIANOV ET AL.with each other ( Tables 3 and 4), suggesting that the inte-\\ngrated computational approach used in this study allowed\\none to avoid false-positive results and correctly estimate thestrength of intermolecular interactions. This assumption is\\nindirectly supported by the data of a newly study of Shen &\\ncoauthors (Shen et al., 2021 ) in which the impact of the com-\\nbinations of features from multiple scoring functions on the\\nscoring power was explored. According to this study, com-\\nbining NNScore 2.0 with one to four other classical scoringfunctions may provide the best accuracy of binding affinity\\nprediction. Altogether, these findings ensure strong evidencesuggesting that the identified molecules may exhibit the low\\nvalues of binding free energy in the complexes with gp120\\nsimilar to those calculated for the HIV-1 entry inhibitors NBD-11021 and NBD-14010. The binding affinity and interaction\\nmodes predicted for the analyzed compounds and gp120\\ntestify to that these small drug-like molecules may serve asgood scaffolds for the development of novel anti-HIV-1\\nagents able to target the gp120 CD4-binding site. However,\\ndespite their hopeful in silico profile, the predicted CD4-\\nmimetic candidates are only promising basic structures to\\ndevelop new potent and broad HIV-1 entry inhibitors. In this\\nFigure 8. The gp120 residues forming van der Waals contacts with compounds ZINC000026430653 (a), ZINC000191389930 (b) and ZINC000293423658 (c).\\nTable 4. Mean values of binding free energy <DG>for the predicted CD4-mimetic candidates, NBD-11021 and NBD-14010\\nbound to gp120 and their standard deviations DGSTD1.\\nCompound DH±DHSTD,kcal/mol T DS±TDSSTD,kcal/mol DG±DGSTD,kcal/mol\\nZINC000026430653 /C040.5 ± 3.1 /C020.3 ± 5.4 /C020.2 ± 6.1\\nZINC000191389930 /C031.3 ± 2.9 /C020.3 ± 6.1 /C011.0 ± 6.9\\nZINC000293423658 /C042.6 ± 3.9 /C020.9 ± 6.2 /C021.7 ± 7.3\\nNBD-11021 /C042.2 ± 3.7 /C022.9 ± 6.9 /C019.3 ± 7.9\\nNBD-14010 /C039.7 ± 2.9 /C021.8 ± 6.1 /C018.0 ± 6.4\\n1<DH>and<TDS>are the mean values of enthalpic and entropic components of free energy respectively; ( DH)STDand\\n(TDS)STDare standard deviations corresponding to these values.JOURNAL OF BIOMOLECULAR STRUCTURE AND DYNAMICS 7567regard, before biomedical testing, the identified compounds\\nshould fall under a lead optimization, iterative process ofaltering the molecule structures to identify their chemical\\nmodifications with improved antiviral activity and acceptable\\nADMET parameters. In doing so, the current QSAR strategiesthat are widely used for a lead optimization approach in\\ndrug discovery may be applied (Golbraikh et al., 2017 ;\\nKuseva et al., 2019 ; Schultz et al., 2018 ). In particular, these\\ntechnologies may be used to design highly potent andbroad analogs of ZINC000026430653, ZINC000191389930\\nand ZINC000293423658 able to form a strong hydrogen\\nbond with Asp-368\\ngp120 instead of a weak C –H/C1/C1/C1O inter-\\naction predicted for these compounds bound to gp120(Figure 7 ).\\nConclusions\\nA generative adversarial autoencoder for anti-HIV-1 drug pre-\\ndiction was developed and applied to identify potential HIV-1 entry inhibitors that target the hydrophobic Phe-43 cavityof gp120 critically important for the virus attachment to the\\nCDþT cells. The validation of the developed autoencoder\\nusing a wide range of chemical compounds from the ZINC15\\ndatabase shows that this neural network combined with vir-tual screening of chemical databases by molecular finger-\\nprints and traditional methods of molecular modeling forms\\na productive platform for discovery of promising anti-HIV-1\\nagents able to target CD4-binding site of the viral envelope\\ngp120 protein. According to the data from molecular dock-ing, machine learning, quantum chemical calculations, and\\nmolecular dynamics simulations, the small molecules\\nZINC000026430653, ZINC000191389930 and\\nZINC000293423658 ( Figure 6 ) exhibit a high-affinity binding\\nto gp120 ( Tables 3 and 4) which is provided by interaction\\nmodes dominating the gp120 /C0CD4 interface ( Figure 7 ). In\\naddition, these molecules fully satisfy the criteria imposed on\\npotential drug by Lipinski ’s‘rule of five ’(Lipinski et al., 2001 ),\\nwhich is confirmed by their physicochemical parameters\\ngiven in the caption for Figure 6 . These drug-like molecules\\nmay be used therefore as good starting points for design of\\nFigure 9. The time dependences of the RMSD (Å) calculated between all of the MD ligand/gp120 structures and the MD models obtained after the 10 ns time\\ndomain. The backbone atoms of gp120 were used in the calculations. The graphs shown correspond to the following ligand/gp120 structures: (a)\\nZINC000026430653/gp120, (b) ZINC000191389930/gp120, (c) ZINC000293423658/gp120, (d) NBD-11021/gp120 and (e) NBD-14010/gp120.7568 A. M. ANDRIANOV ET AL.novel potent antiviral drugs inhibiting the early stages of\\nHIV-1 infection. In this context, the further advancement of\\nthe current study proposes the use of these CD4-mimetic\\ncandidates as the basic structures for computer-assisted gen-eration of their chemical modifications with the higher anti-HIV activity and improved ADMET parameters followed by\\nsynthesis and detailed biochemical assays.\\nWhen analyzing the obtained findings, it is necessary to\\nkeep in mind that only five MACCS structural keys generatedby the neural network for each of two threshold energy val-\\nues were used in this study to validate the operation of thedeveloped autoencoder, which resulted in the identification\\nof three compounds exhibiting strong attachment to the\\nhydrophobic Phe-43 cavity of gp120. However, this neuralnetwork can generate a lot of molecular fingerprints at a\\ngiven maximum allowable value of binding free energy. It is\\nclear that the use of a wider set of these chemical descrip-tors should provide the discovery of novel high-affinity com-\\npounds as promising CD4-mimetic candidates for the design\\nof potent, broad and safe antiviral therapeutics. In this con-nection, the further advancement of this work assumes a fin-\\ngerprint-based screening of different chemical databases\\nusing more MACCS structural keys constructed by the pro-posed neural network.\\nDisclosure statement\\nNo potential conflict of interest was reported by the authors.\\nFunding\\nThis study was supported by a grant of the State Program of Scientific\\nResearch ‘Convergence 2020 ’(subprogram ‘Consolidation ’, project 3.08)\\nand the Belarusian Republican Foundation for Fundamental Research\\n(projects X20 /C140/C145-006 and F21COVID-002).\\nReferences\\nAcharya, P., Lusvarghi, S., Bewley, C. A., & Kwong, P. D. (2015). HIV-1\\ngp120 as a therapeutic target: Navigating a moving labyrinth. Expert\\nOpinion on Therapeutic Targets ,19(6), 765 –783. https://doi.org/10.\\n1517/14728222.2015.1010513\\nAgastheeswaramoorthy, K., & Sevilimedu, A. (2020). Drug REpurposing\\nusing AI/ML tools –for Rare Diseases (DREAM-RD): A case study with\\nFragile X Syndrome (FXS). bioRxiv. https://doi.org/10.1101/2020.09.25.\\n311142\\nAin, Q. U., Aleksandrova, A., Roessler, F. D., & Ballester, P. J. (2015).\\nMachine-learning scoring functions to improve structure-based bind-\\ning affinity prediction and virtual screening. Wiley Interdisciplinary\\nReviews. Computational Molecular Science ,5(6), 405 –424. https://doi.\\norg/10.1002/wcms.1225\\nAlhossary, A., Handoko, S. D., Mu, Y., & Kwoh, C. K. (2015). Fast, accurate,\\nand reliable molecular docking with QuickVina 2. Bioinformatics ,\\n31(13), 2214 –2216. https://doi.org/10.1093/bioinformatics/btv082\\nAndrianov, A. M., Kashyn, I. A., & Tuzikov, A. V. (2017). Computational\\nidentification of novel entry inhibitor scaffolds mimicking primary\\nreceptor CD4 of HIV-1 gp120. Journal of Molecular Modeling ,23(1), 18.\\nhttps://doi.org/10.1007/s00894-016-3189-4\\nAndrianov, A. M., Nikolaev, G. I., Kornoushenko, Y. V., Xu, W., Jiang, S., &\\nTuzikov, A. V. (2019). In silico identification of novel aromatic com-pounds as potential HIV-1 entry inhibitors mimicking cellular receptor\\nCD4. Viruses ,11(8), 746. https://doi.org/10.3390/v11080746\\nAntoine, D., Olivier, M., Vincent, Z. (2017). Swiss ADME: A free web tool\\nto evaluate pharmacokinetics, drug-likeness and medicinal chemistry\\nfriendliness of small molecules. Scientific Reports ,7, 42717. https://doi.\\norg/10.1038/srep42717\\nArts, E. J., & Hazuda, D. J. (2012). HIV-1 antiretroviral drug therapy. Cold\\nSpring Harbor Perspectives in Medicine ,2(4), a007161. https://doi.org/\\n10.1101/cshperspect.a007161\\nAr/C19us-Pous, J., Blaschke, T., Ulander, S., Reymond, J. L., Chen, H., &\\nEngkvist, O. (2019). Exploring the GDB-13 chemical space using deepTable 5. Averages of the binding enthalpy for the amino-acid residues of\\ngp120 bound to the predicted compounds, NBD-11021 and NBD-140101,2,3,4,5.\\nCompound\\nResidue of gp120 I II III NBD-14010 NBD-11021\\nResidue Contribution to the Binding Enthalpy (kcal/mol)\\nVal-255 -0.9 ± 0.7 -0.6 ± 0.2 -0.5 ± 0.3 -0.6 ± 0.3 -1.1 ± 0.3\\nSer-256 /C00.5 ± 0.3 – /C00.6 ± 0.2 – /C00.5 ± 0.2\\nThr-257 -1.4 ± 0.4 -1.2 ± 0.4 -1.3 ± 0.3 -1.2 ± 0.2 -0.5 ± 0.4Pro-369 –––– /C00.7 ± 0.3\\nGlu-370 -0.9 ± 0.6 -0.6 ± 0.8 -1.3 ± 0.7 -1.4 ± 0.4 -1.5 ± 0.5\\nIle-371 -0.8 ± 0.4 -1.2 ± 1.0 -1.1 ± 0.5 -1.2 ± 0.5 -1.0 ± 0.5Ser-375 /C00.5 ± 0.3 /C00.5 ± 0.3 /C00.6 ± 0.4 /C00.7 ± 0.2 –\\nPhe-376 –– /C00.6 ± 0.2 – /C00.5 ± 0.2\\nPhe-382 /C01.0 ± 0.6 – /C00.8 ± 0.3 /C00.5 ± 0.1 /C00.6 ± 0.3\\nTyr-384 –––– /C00.7 ± 0.3\\nIle-424 /C00.5 ± 0.3 – /C01.3 ± 0.4 /C00.7 ± 0.2 /C00.8 ± 0.3\\nAsn-425 – /C00.8 ± 0.5 –– /C00.7 ± 0.4\\nTrp-427 -2.2 ± 0.6 -1.6 ± 0.4 -1.8 ± 0.5 -2.4 ± 0.5 -2.0 ± 0.4\\nVal-430 ––– /C00.9 ± 0.2 0.5 ± 0.6\\nGly-473 – /C00.8 ± 0.4 – /C00.5 ± 0.4 –\\nAsp-474 /C00.5 ± 0.4 –– /C00.5 ± 0.5 /C00.7 ± 0.3\\nMet-475 -1.4 ± 0.5 -0.8 ± 0.4 -1.3 ± 0.4 -0.8 ± 0.3 -0.7 ± 0.3\\nLigand\\n4/C026.6 ± 2.1 /C020.8 ± 1.5 /C026.3 ± 1.6 /C027.4 ± 1.6 /C029.2 ± 1.9\\n1Data for the gp120 residues with the binding enthalpy /C20-0.4 kcal/mol\\nare presented.\\n2The averages of the residue contributions to the binding enthalpy and corre-\\nsponding standard deviations are given.\\n3The gp120 residues dominating the ligand/gp120 interaction are highlighted\\nby bold.\\n4The ligand contributions to the binding enthalpy are presented.\\n5Compounds ZINC000026430653, ZINC000191389930 and ZINC000293423658are designated by numerals I, II and III, respectively.\\nTable 6. Values of RMSF for the gp120 residues contributing to the bind-\\ning enthalpy1,2.\\nCompound\\nResidue I II III NBD-14010 NBD-11021\\nValues of RMSF (Å) for the individual residues of gp120\\nVal-255 0.80 0.67 1.33 0.56 0.78\\nSer-256 0.73 – 0.98 – 0.71\\nThr-257 0.66 0.65 0.73 0.50 0.71Pro-369 ––– – 1.46\\nGlu-370 0.94 1.56 1.24 0.94 1.12\\nIle-371 1.08 1.21 1.17 0.92 1.00Ser-375 0.65 0.81 0.72 0.69 –\\nPhe-376 –– 0.77 – 1.06\\nPhe-382 0.64 – 0.66 0.57 1.14\\nTyr-384 ––– – 0.99\\nIle-424 1.02 – 0.90 0.73 1.36\\nAsn-425 – 1.11 –– 1.31\\nTrp-427 1.15 1.28 0.92 0.73 1.13\\nVal-430 ––– 0.87 1.57\\nGly-473 – 1.31 – 1.42 –\\nAsp-474 0.84 –– 1.03 0.81\\nMet-475 0.73 0.80 0.79 0.83 0.75\\n1The gp120 residues presenting the dominant contributors to the ligand/\\ngp120 interaction are highlighted by bold.\\n2Compounds ZINC000026430653, ZINC000191389930 and ZINC000293423658are designated by numerals I, II and III, respectively.JOURNAL OF BIOMOLECULAR STRUCTURE AND DYNAMICS 7569generative models. Journal of Cheminformatics ,11(1), 20. https://doi.\\norg/10.1186/s13321-019-0341-z\\nAr/C19us-Pous, J., Johansson, S. V., Prykhodko, O., Bjerrum, E. J., Tyrchan, C.,\\nReymond, J. L., Chen, H., & Engkvist, O. (2019). Randomized SMILESstrings improve the quality of molecular generative models. Journal\\nof Cheminformatics ,11(1), 71. https://doi.org/10.1186/s13321-019-\\n0393-0\\nBallester, P. J., & Mitchell, J. B. (2010). A machine learning approach to\\npredicting protein-ligand binding affinity with applications to molecu-lar docking. Bioinformatics ,26(9), 1169 –1175. https://doi.org/10.1093/\\nbioinformatics/btq112\\nBian, Y., Jing, Y., Wang, L., Ma, S., Jun, J. J., & Xie, X.-Q. (2019). Prediction\\nof orthosteric and allosteric regulations on cannabinoid receptors\\nusing supervised machine learning classifiers. Molecular\\nPharmaceutics ,16(6), 2605 –2615. https://doi.org/10.1021/acs.molphar-\\nmaceut.9b00182\\nBlaschke, T., Ar /C19us-Pous, J., Chen, H., Margreitter, C., Tyrchan, C., Engkvist,\\nO., Papadopoulos, K., & Patronov, A. (2020). REINVENT 2.0 –An AI tool\\nfor de novo drug design. Journal of Chemical Information and\\nModeling ,60(12), 5918 –5922. https://doi.org/10.1021/acs.jcim.0c00915\\nBrylinski, M. (2013). Nonlinear scoring functions for similarity-based lig-\\nand docking and binding affinity prediction. Journal of Chemical\\nInformation and Modeling ,53(11), 3097 –3112. https://doi.org/10.1021/\\nci400510e\\nCase, D. A., Belfon, K., Ben-Shalom, I. Y., Brozell, S. R., Cerutti, D. S.,\\nCheatham, T. E., III, & Kollman, P. A. (2020). AMBER 2020 . University of\\nCalifornia.\\nCavasotto, C. N., Adler, N. S., & Aucar, M. G. (2018). Quantum chemical\\napproaches in structure-based virtual screening and lead optimiza-tion. Frontiers in Chemistry ,6, 188. https://doi.org/10.3389/fchem.2018.\\n00188\\nChen, H., Engkvist, O., Wang, Y., Olivecrona, M., & Blaschke, T. (2018). The\\nrise of deep learning in drug discovery. Drug Discovery Today ,23(6),\\n1241 –1250. https://doi.org/10.1016/j.drudis.2018.01.039\\nCherkasov, A., Muratov, E. N., Fourches, D., Varnek, A., Baskin, I. I., Cronin,\\nM., Dearden, J., Gramatica, P., Martin, Y. C., Todeschini, R., Consonni,\\nV., Kuz ’min, V. E., Cramer, R., Benigni, R., Yang, C., Rathman, J.,\\nTerfloth, L., Gasteiger, J., Richard, A., & Tropsha, A. (2014). QSAR mod-eling: Where have you been? Where are you going to? Journal of\\nMedicinal Chemistry ,57(12), 4977 –\\n5010. https://doi.org/10.1021/\\njm4004285\\nCreswell, A., White, T., Dumoulin, V., Arulkumaran, K., Sengupta, B., &\\nBharath, A. A. (2018). Generative adversarial networks: An overview.\\nIEEE Signal Processing Magazine ,35(1), 53 –65.https://doi.org/10.1109/\\nMSP.2017.2765202\\nCurreli, F., Kwon, Y. D., Zhang, H., Scacalossi, D., Belov, D. S., Tikhonov,\\nA. A., Andreev, I. A., Altieri, A., Kurkin, A. V., Kwong, P. D., & Debnath,A. K. (2015). Structure-based design of a small molecule CD4-antagon-ist with broad spectrum anti-HIV-1 activity. Journal of Medicinal\\nChemistry ,58(17), 6909 –6927. https://doi.org/10.1021/acs.jmedchem.\\n5b00709\\nCurreli, F., Belov, D. S., Kwon, Y. D., Ramesh, R., Furimsky, A. M.,\\nO’Loughlin, K., Byrge, P. C., Iyer, L. V., Mirsalis, J. C., Kurkin, A. V.,\\nAltieri, A., & Debnath, A. K. (2018). Structure-based lead optimizationto improve antiviral potency and ADMET properties of phenyl-1H-pyr-role-carboxamide entry inhibitors targeted to HIV-1 gp120. European\\nJournal of Medicinal Chemistry ,154, 367 –391. https://doi.org/10.1016/j.\\nejmech.2018.04.062\\nCurreli, F., Kwon, Y. D., Belov, D. S., Ramesh, R. R., Kurkin, A. V., Altieri, A.,\\nKwong, P. D., & Debnath, A. K. (2017). Synthesis, antiviral potency,in vitro ADMET, and X-ray structure of potent CD4 mimics as entryinhibitors that target the Phe43 cavity of HIV-1 gp120. Journal of\\nMedicinal Chemistry ,60(7), 3124 –3153. https://doi.org/10.1021/acs.\\njmedchem.7b00179\\nCurreli, F., Shahad, A., Benedict, V. S. M., Iusupov, I. R., Belov, D. S.,\\nMarkov, P. O., Kurkin, A. V., Andrea, A., & Debnath, A. K. (2020).Preclinical optimization of gp120 entry antagonists as anti-HIV-1agents with improved cytotoxicity and ADME properties throughrational design, synthesis, and antiviral evaluation. Journal ofMedicinal Chemistry ,63(4), 1724 –1749. https://doi.org/10.1021/acs.\\njmedchem.9b02149\\nDe Clercq, E. (2005). New approaches toward anti-HIV chemotherapy.\\nJournal of Medicinal Chemistry ,48(5), 1297 –1313. https://doi.org/10.\\n1021/jm040158k\\nDesiraju, G. R., & Steiner, T. (1999). The weak hydrogen bond in structural\\nchemistry and biology (p. 507). Oxford University Press.\\nDobchev, D. A., Pillai, G. G., & Karelson, M. (2014). In silico machine learn-\\ning methods in drug development. Current Topics in Medicinal\\nChemistry , 14(16), 1913 –1922. https://doi.org/10.2174/\\n1568026614666140929124203\\nDubey, A. (2018). Machine learning approaches in drug development of\\nHIV/AIDS. International Journal of Molecular Biology ,3(1), 23 –25.\\nhttps://doi.org/10.15406/ijmboa.2018.03.00044\\nDumoulin, V., Belghazi, I., Poole, B., Mastropietro, O., Lamb, A., Arjovsky,\\nM., & Courville, A. (2017). Adversarially learned inference. arXiv:1606.00704v3.\\nDurrant, J. D., & McCammon, J. A. (2011a). NNScore 2.0: A neural-net-\\nwork receptor-ligand scoring function. Journal of Chemical Information\\nand Modeling ,51\\n(11), 2897 –2903. https://doi.org/10.1021/ci2003889\\nDurrant, J. D., & McCammon, J. A. (2011b). BINANA: A novel algorithm\\nfor ligand-binding characterization. Journal of Molecular Graphics &\\nModelling ,29(6), 888 –893. https://doi.org/10.1016/j.jmgm.2011.01.004\\nDurrant, J. D., & McCammon, J. A. (2012). AutoClickChem: Click chemistry\\nin silico. PLoS Computational Biology ,8(3), e1002397. https://doi.org/\\n10.1371/journal.pcbi.1002397\\nDuvenaud, D., Maclaurin, D., Aguilera-Iparraguirre, J., G /C19omez-Bombarelli,\\nR., Hirzel, T., Aspuru-Guzik, A., & Adams, R. P. (2015). Convolutional\\nnetworks on graphs for learning molecular fingerprints. arXiv:\\n1509.09292.\\nEkins, S. (2016). The next era: Deep learning in pharmaceutical Research.\\nPharmaceutical Research ,33(11), 2594 –2603. https://doi.org/10.1007/\\ns11095-016-2029-7\\nEssmann, U., Perera, L., Berkowitz, M. L., Darden, T., Lee, H., & Pedersen,\\nL. G. (1995). A smooth particle mesh Ewald method. The Journal of\\nChemical Physics ,103(19), 8577 –8593. https://doi.org/10.1063/1.470117\\nEste, J. A., & Telenti, A. (2007). HIV entry inhibitors. The Lancet ,370(9581),\\n81–88.https://doi.org/10.1016/S0140-6736(07)61052-6\\nGenheden, S., & Ryde, U. (2015). The MM/PBSA and MM/GBSA methods\\nto estimate ligand-binding affinities. Expert Opinion on Drug Discovery ,\\n10(5), 449 –461. https://doi.org/10.1517/17460441.2015.1032936\\nGibbs, M. N., & MacKay, D. C. (2000). Variational Gaussian process classi-\\nfiers. IEEE Transactions on Neural Networks ,11(6), 1458 –1464. https://\\ndoi.org/10.1109/72.883477\\nGolbraikh, A., Wang, X. S., Zhu, H., & Tropsha, A. (2017). Predictive QSAR\\nmodeling: Methods and applications in drug discovery and chemicalrisk assessment. In J. Leszczynski, A. Kaczmarek-Kedziera, T. Puzyn,G. M. Papadopoulos, H. Reis, & K. M. Shukla (Eds.), Handbook of com-\\nputational chemistry (pp. 2303 –2340). Springer.\\nG/C19omez-Bombarelli, R., Wei, J. N., Duvenaud, D., Hern /C19andez-Lobato, J. M.,\\nS/C19anchez-Lengeling, B., Sheberla, D., Aguilera-Iparraguirre, J., Hirzel,\\nT. D., Adams, R. P., & Aspuru-Guzik, A. (2018). Automatic chemicaldesign using a data-driven continuous representation of molecules.ACS Central Science ,4(2), 268 –276. https://doi.org/10.1021/acscentsci.\\n7b00572\\nGoodfellow, I. J., Pouget-Abadie, J., Mirza, M., Xu, B., Warde-Farley, D.,\\nOzair, S., Courville, A., & Bengio, Y. (2014). Generative adversarial net-works. arXiv:1406.2661.\\nGuedes, I. A., Pereira, F. S. S., & Dardenne, L. E. (2018). Empirical scoring\\nfunctions for structure-based virtual screening: Applications, critical\\naspects, and challenges. Frontiers in Pharmacology ,9, 1089. https://\\ndoi.org/10.3389/fphar.2018.01089\\nGuimaraes, G. L., Sanchez-Lengeling, B., Outeiral, C., Farias, P. L. C., &\\nAspuru-Guzik, A. (2017). Objective-reinforced generative adversarialnetworks (ORGAN) for sequence generation models. ArXiv preprintarXiv:1705.10843.\\nGupta, A., Muller, A. T., Huisman, B. J. H., Fuchs, J. A., Schneider, P., &\\nSchneider, G. (2018). Generative recurrent networks for de novo drugdesign. Molecular Informatics ,37(1-2), 1700111. https://doi.org/10.\\n1002/minf.2017001117570 A. M. ANDRIANOV ET AL.Hamming, R. W. (1950). Error detecting and error correcting codes. Bell\\nSystem Technical Journal ,29(2), 147 –160. https://doi.org/10.1002/j.\\n1538-7305.1950.tb00463.x\\nHein, C. D., Liu, X.-M., & Wang, D. (2008). Click chemistry, a powerful tool\\nfor pharmaceutical sciences. Pharmaceutical Research ,25(10),\\n2216 –2230. https://doi.org/10.1007/s11095-008-9616-1\\nHollingsworth, S. A., & Dror, R. O. (2018). Molecular dynamics simulation\\nfor all. Neuron ,99(6), 1129 –1143. https://doi.org/10.1016/j.neuron.\\n2018.08.011\\nHøyvik, I.-M., Jansik, B., & Jørgensen, P. (2012). Trust region minimization\\nof orbital localization functions. Journal of Chemical Theory and\\nComputation ,8(9), 3137 –3146. https://doi.org/10.1021/ct300473g\\nJaeger, S., Fulle, S., & Turk, S. (2018). Mol2vec: Unsupervised machine\\nlearning approach with chemical intuition. Journal of Chemical\\nInformation and Modeling ,58(1), 27 –35. https://doi.org/10.1021/acs.\\njcim.7b00616\\nJanocha, K., & Czarnecki, W. M. (2017). On loss functions for deep neural\\nnetworks in classification. Schedae Informaticae ,1/2016 ,4 9–59.https://\\ndoi.org/10.4467/20838476SI.16.004.6185\\nJimenez-Luna, J., Grisoni, F., & Schneider, G. (2020). Drug discovery with\\nexplainable artificial intelligence. Nature Machine Intelligence ,2(10),\\n573–584. https://doi.org/10.1038/s42256-020-00236-4\\nJorgensen, W. L., Chandrasekhar, J., Madura, J. D., Impey, R. W., & Klein,\\nM. L. (1983). Comparison of simple potential functions for simulatingliquid water. The Journal of Chemical Physics ,79(2), 926 –935. https://\\ndoi.org/10.1063/1.445869\\nKadurin, A., Aliper, A., Kazennov, A., Mamoshina, P., Vanhaelen, Q.,\\nKhrabrov, K., & Zhavoronkov, A. (2017). The cornucopia of meaningful\\nleads: Applying deep adversarial autoencoders for new molecule\\ndevelopment in oncology. Oncotarget ,8(7), 10883 –10890. https://doi.\\norg/10.18632/oncotarget.14073\\nKadurin, A., Nikolenko, S., Khrabrov, K., Aliper, A., & Zhavoronkov, A.\\n(2017). DruGAN: An advanced generative adversarial autoencodermodel for de novo generation of new molecules with desiredmolecular properties in silico. Molecular Pharmaceutics ,14(9),\\n3098 –\\n3104. https://doi.org/10.1021/acs.molpharmaceut.7b00346\\nKatsila, T., Spyroulias, G. A., Patrinos, G. P., & Matsoukas, M.-T. (2016).\\nComputational approaches in target identification and drug discovery.Computational and Structural Biotechnology Journal ,14, 177 –184.\\nhttps://doi.org/10.1016/j.csbj.2016.04.004\\nKerns, E. H., & Di, L. (2008). Drug-like properties: Concepts, structure, design\\nand methods: From ADME to toxicity optimization . Elsevier Inc.\\nKingma, D. P., & Ba, J. (2014). Adam: A method for stochastic optimiza-\\ntion. arXiv: preprint arXiv: 1412.6980.\\nKinnings, S. L., Liu, N., Tonge, P. J., Jackson, R. M., Xie, L., & Bourne, P. E.\\n(2011). A Machine learning-based method to improve docking scoringfunctions and its application to drug repurposing. Journal of Chemical\\nInformation and Modeling ,51(2), 408 –419. https://doi.org/10.1021/\\nci100369f\\nKlamt, A. (2005). COSMO-RS: From quantum chemistry to fluid phase ther-\\nmodynamics and drug design (1st ed., p. 246). Elsevier.\\nKlamt, A., & Sch €u€urmann, G. (1993). COSMO: A new approach to dielec-\\ntric screening in solvents with explicit expressions for the screeningenergy and its gradient. Journal of the Chemical Society, Perkin\\nTransactions ,2, 799 –805.\\nKnight-Schrijver, V. R., Chelliah, V., Cucurull-Sanchez, L., & Le Nov /C18ere, N.\\n(2016). The promises of quantitative systems pharmacology modellingfor drug development. Computational and Structural Biotechnology\\nJournal ,14, 363 –370. https://doi.org/10.1016/j.csbj.2016.09.002\\nKolb, H. C., Finn, M. G., & Sharpless, K. B. (2001). Click chemistry: Diverse\\nchemical function from a few good reactions. Angewandte Chemie\\nInternational Edition ,40 (11), 2004 –2021. doi.org/10.1002/1521-\\n3773(20010601)40:11 https://doi.org/10.1002/1521-3773(20010601)40:\\n11<2004::AID-ANIE2004 >3.0.CO;2-5\\nKumari, G., & Singh, R. K. (2012). Highly active antiretroviral therapy for\\ntreatment of HIV/AIDS patients: Current status and future prospectsand the Indian scenario. HIV & AIDS Review ,11(1), 5 –14.https://doi.\\norg/10.1016/j.hivar.2012.02.003\\nKuseva, C., Schultz, T. W., Yordanova, D., Tankova, K., Kutsarova, S.,\\nPavlov, T., Chapkanov, A., Georgiev, M., Gissi, A., Sobanski, T., &Mekenyan, O. G. (2019). The implementation of RAAF in the OECD\\nQSAR Toolbox. Regulatory Toxicology and Pharmacology ,105,5 1–61.\\nhttps://doi.org/10.1016/j.yrtph.2019.03.018\\nKwong, P. D., Wyatt, R., Robinson, J., Sweet, R. W., Sodroski, J., &\\nHendrickson, W. A. (1998). Structure of an HIV gp120 envelope glyco-\\nprotein in complex with the CD4 receptor and a neutralizing human\\nantibody. Nature ,393(6686), 648\\n–659. https://doi.org/10.1038/31405\\nLavecchia, A. (2019). Deep learning in drug discovery: Opportunities,\\nchallenges and future prospects. Drug Discovery Today ,24(10),\\n2017 –2032. https://doi.org/10.1016/j.drudis.2019.07.006\\nLeelananda, S. P., & Lindert, S. (2016). Computational methods in drug\\ndiscovery. Beilstein Journal of Organic Chemistry ,12, 2694 –2718.\\nhttps://doi.org/10.3762/bjoc.12.267\\nLi, H., Sze, K.-H., Lu, G., & Ballester, P. J. (2021). Machine-learning scoring\\nfunctions for structure-based virtual screening. WIREs Computational\\nMolecular Science ,11(1), e1478. https://doi.org/10.1002/wcms.1478\\nLi, W., Lu, L., Li, W., & Jiang, S. (2017). Small-molecule HIV-1 entry inhibi-\\ntors targeting gp120 and gp41: A patent review (2010-2015). Expert\\nOpinion on Therapeutic Patents ,27(6), 707 –719. https://doi.org/10.\\n1080/13543776.2017.1281249\\nLipinski, C. A., Lombardo, F., Dominy, B. W., & Feeney, P. J. (2001).\\nExperimental and computational approaches to estimate solubility\\nand permeability in drug discovery and development settings.\\nAdvanced Drug Delivery Reviews ,46(1-3), 3 –26. PMID: 11259830\\nhttps://doi.org/10.1016/s0169-409x(00)00129-0\\nLipinski, C. F., Maltarollo, V. G., Oliveira, P. R., da Silva, A. B. F., & Honorio,\\nK. M. (2019). Advances and perspectives in applying deep learning for\\ndrug design and discovery. Frontiers in Robotics and AI ,6, 108. https://\\ndoi.org/10.3389/frobt.2019.00108\\nMacArthur, R. D., & Novak, R. M. (2008). Reviews of anti-infective agents:\\nMaraviroc: The first of a new class of antiretroviral agents. Clinical\\nInfectious Diseases ,47(2), 236 –241. https://doi.org/10.1086/589289\\nMallipeddi, P. L., Kumar, G., White, S. W., & Webb, T. R. (2014). Recent\\nadvances in computer-aided drug design as applied to anti-influenza\\ndrug discovery. Current Topics in Medicinal Chemistry ,14(16),\\n1875 –1889. https://doi.org/10.2174/1568026614666140929153812\\nMatthews, T., Salgo, M., Greenberg, M., Chung, J., DeMasi, R., &\\nBolognesi, D. (2004). Enfuvirtide: The first therapy to inhibit the entry\\nof HIV-1 into host CD4 lymphocytes. Nature Reviews. Drug Discovery ,3\\n(3), 215 –225. https://doi.org/10.1038/nrd1331\\nMcDonald, I. K., & Thornton, J. M. (1994). Satisfying hydrogen bonding\\npotential in proteins. Journal of Molecular Biology ,238 (5), 777 –793.\\nhttps://doi.org/10.1006/jmbi.1994.1334\\nM/C19endez-Lucio, O., Baillif, B., Clevert, D.-A., Rouqui /C19e, D., & Wichard, J.\\n(2020). De novo generation of hit-like molecules from gene expres-\\nsion signatures using artificial intelligence. Nature Communications ,\\n11(1), 10. https://doi.org/10.1038/s41467-019-13807-w\\nMeng, X.-Y., Zhang, X.-H., Mezei, M., & Cui, M. (2011). Molecular Docking:\\nA powerful approach for structure-based drug discovery. Current\\nComputer-Aided Drug Design ,7(2), 146 –157. https://doi.org/10.2174/\\n157340911795677602\\nMercado, R., Rastemo, T., Lindel €of, E., Klambauer, G., Engkvist, O., Chen,\\nH., & Bjerrum, E. J. (2020). Practical notes on building molecular graph\\ngenerative models. ChemRxiv. Preprint. https://doi.org/10.26434/\\nchemrxiv.12888383\\nMoses, J. E., & Moorhouse, A. D. (2007). The growing applications of click\\nchemistry. Chemical Society Reviews ,36(8), 1249 –1262. https://doi.org/\\n10.1039/B613014N\\nMyszka, D. G., Sweet, R. W., Hensley, P., Brigham-Burke, M., Kwong, P. D.,\\nHendrickson, W. A., Wyatt, R., Sodroski, J., & Doyle, M. L. (2000).\\nEnergetics of the HIV gp120-CD4 binding reaction. Proceedings of the\\nNational Academy of Sciences of the United States of America ,97(16),\\n9026 –9031. https://doi.org/10.1073/pnas.97.16.9026\\nOlivecrona, M., Blaschke, T., Engkvist, O., & Chen, H. (2017). Molecular de-\\nnovo design through deep reinforcement learning. Journal of\\nCheminformatics ,9(1), 48. https://doi.org/10.1186/s13321-017-0235-x\\nPettersen, E. F., Goddard, T. D., Huang, C. C., Couch, G. S., Greenblatt,\\nD. M., Meng, E. C., & Ferrin, T. E. (2004). UCSF Chimera –A visualiza-\\ntion system for exploratory research and analysis. Journal ofJOURNAL OF BIOMOLECULAR STRUCTURE AND DYNAMICS 7571Computational Chemistry ,25(13), 1605 –1612. https://doi.org/10.1002/\\njcc.20084\\nPolykovskiy, D., Zhebrak, A., Vetrov, D., Ivanenkov, Y., Aladinskiy, V.,\\nMamoshina, P., Bozdaganyan, M., Aliper, A., Zhavoronkov, A., &\\nKadurin, A. (2018). Entangled conditional adversarial autoencoder for\\nde novo drug discovery. Molecular Pharmaceutics ,15(10), 4398 –4405.\\nhttps://doi.org/10.1021/acs.molpharmaceut.8b00839s\\nPrykhodko, O., Johansson, S. V., Kotsias, P. C., Ar /C19us-Pous, J., Bjerrum, E. J.,\\nEngkvist, O., & Chen, H. (2019). A de novo molecular generationmethod using latent vector based generative adversarial network.Journal of Cheminformatics ,11(1), 74. https://doi.org/10.1186/s13321-\\n019-0397-9\\nQureshi, A., Rajput, A., Kaur, G., & Kumar, M. (2018). HIVprotI: An inte-\\ngrated web based platform for prediction and design of HIV proteins\\ninhibitors. Journal of Cheminformatics ,10(1), 12. https://doi.org/10.\\n1186/s13321-018-0266-y\\nRagoza, M., Hochuli, J., Idrobo, E., Sunseri, J., & Koes, D. R. (2017).\\nProtein-ligand scoring with convolutional neural networks. Journal of\\nChemical Information and Modeling ,57(4), 942 –957. https://doi.org/10.\\n1021/acs.jcim.6b00740\\nRappe, A. K., Casewit, C. J., Colwell, K. S., Goddard, W. A., III., & Skiff,\\nW. M. (1992). UFF, a full periodic table force field for molecularmechanics and molecular dynamics simulations. Journal of the\\nAmerican Chemical Society ,114(25), 10024 –10035. https://doi.org/10.\\n1021/ja00051a040\\nRusconi, S., Scozzafava, A., Mastrolorenzo, A., & Supuran, C. T. (2007). An\\nupdate in the development of HIV entry inhibitors. Current Topics in\\nMedicinal Chemistry ,7(13), 1273 –1289. https://doi.org/10.2174/\\n156802607781212239\\nRyckaert, J. P., Ciccotti, G., & Berendsen, H. J. C. (1977). Numerical inte-\\ngration of the Cartesian equations of motion of a system with con-\\nstraints: Molecular dynamics of n-alkanes. Journal of Computational\\nPhysics ,23(3), 327 –341. https://doi.org/10.1016/0021-9991(77)90098-5\\nRyde, U., & S €oderhjelm, P. (2016). Ligand-binding affinity estimates sup-\\nported by quantum-mechanical methods. Chemical Reviews ,116(9),\\n5520 –5566. https://doi.org/10.1021/acs.chemrev.5b00630\\nRyser, H. J.-P., & Fluckiger, R. (2005). Progress in targeting HIV-1 entry.\\nDrug Discovery Today ,10(16), 1085 –1094. https://doi.org/10.1016/\\nS1359-6446(05)03550-6\\nSakano, T., Mahamood, M. I., Yamashita, T., & Fujitani, H. (2016).\\nMolecular dynamics analysis to evaluate docking pose prediction.\\nBiophysics and Physicobiology ,13, 181 –194. https://doi.org/10.2142/\\nbiophysico.13.0_181\\nSalmaso, V., & Moro, S. (2018). Bridging molecular docking to molecular\\ndynamics in exploring ligand-protein recognition process: An over-view. Frontiers in Pharmacology ,9, 923. https://doi.org/10.3389/fphar.\\n2018.00923\\nSanchez-Lengeling, B., Outeiral, C., Guimaraes, G. L., & Aspuru-Guzik, A.\\n(2017). Optimizing distributions over molecular space. An Objective-Reinforced Generative Adversarial Network for Inverse-Design\\nChemistry (ORGANIC). Preprint from ChemRxiv .https://doi.org/10.\\n26434/chemrxiv.5309668.v3 .\\nSander, T., Freyss, J., von Korff, M., & Rufener, C. (2015). DataWarrior: An\\nopen-source program for chemistry aware data visualization and ana-\\nlysis. Journal of Chemical Information and Modeling ,55(2), 460 –473.\\nhttps://doi.org/10.1021/ci500588j\\nSchultz, T. W., Diderich, R., Kuseva, C. D., & Mekenyan, O. G. (2018). The\\nOECD QSAR Toolbox starts its second decade. Methods in Molecular\\nBiology ,1800 ,5 5–77.https://doi.org/10.1007/978-1-4939-7899-1_2\\nSegler, M. H., Kogej, T., Tyrchan, C., & Waller, M. P. (2018). Generating\\nfocused molecule libraries for drug discovery with recurrent neuralnetworks. ACS Central Science ,4(1), 120 –131. https://doi.org/10.1021/\\nacscentsci.7b00512\\nSenior, A. W., Evans, R., Jumper, J., Kirkpatrick, J., Sifre, L., Green, T., Qin,\\nC.,/C20Z/C19ıdek, A., Nelson, A. W. R., Bridgland, A., Penedones, H., Petersen,\\nS., Simonyan, K., Crossan, S., Kohli, P., Jones, D. T., Silver, D.,\\nKavukcuoglu, K., & Hassabis, D. (2020). Improved protein structure\\nprediction using potentials from deep learning. Nature ,577, 706 –710.\\nhttps://doi.org/10.1038/s41586-019-1923-7Sharma, A., & Lal, S. P. (2011). Tanimoto based similarity measure for\\nintrusion detection system. Journal of Information Security ,02(04),\\n195–201. https://doi.org/10.4236/jis.2011.24019\\nSharma, G., & First, E. A. (2009). Thermodynamic analysis reveals a tem-\\nperature-dependent change in the catalytic mechanism of bacillus\\nstearothermophilus tyrosyl-tRNA synthetase. The Journal of Biological\\nChemistry ,284(7), 4179 –4190. https://doi.org/10.1074/jbc.M808500200\\nShen, C., Hu, Y., Wang, Z., Zhong, H., Zhang, H., Zhong, H., Wang, G.,\\nYao, X., Xu, L., Cao, D., & Hou, T. (2021). Can machine learning consist-\\nently improve the scoring power of classical scoring functions?Insights into the role of machine learning in scoring functions.Briefings in Bioinformatics ,22(1), 497\\n–514. https://doi.org/10.1093/bib/\\nbbz173\\nSilver, D., Huang, A., Maddison, C. J., Guez, A., Sifre, L., van den\\nDriessche, G., Schrittwieser, J., Antonoglou, I., Panneershelvam, V.,Lanctot, M., Dieleman, S., Grewe, D., Nham, J., Kalchbrenner, N.,Sutskever, I., Lillicrap, T., Leach, M., Kavukcuoglu, K., Graepel, T., &Hassabis, D. (2016). Mastering the game of go with deep neural net-works and tree search. Nature ,529(7587), 484 –489. https://doi.org/10.\\n1038/nature16961\\nSliwoski, G., Kothiwale, S., Meiler, J., & Lowe, EWJr. (2014). Computational\\nmethods in drug discovery. Pharmacological Reviews ,66(1), 334 –395.\\nhttps://doi.org/10.1124/pr.112.007336\\nSteiner, T. (2003). C –H/C15/C15/C15O hydrogen bonding in crystals.\\nCrystallography Reviews ,9(2-3), 177 –228. https://doi.org/10.1080/\\n08893110310001621772\\nSterling, T., & Irwin, J. J. (2015). ZINC 15-ligand discovery for everyone.\\nJournal of Chemical Information and Modeling ,55(11), 2324 –2337.\\nhttps://doi.org/10.1021/acs.jcim.5b00559\\nStewart, J. J. P. (2013). Optimization of parameters for semiempirical\\nmethods VI: More modifications to the NDDO approximations and re-\\noptimization of parameters. Journal of Molecular Modeling ,19(1),\\n1–32.https://doi.org/10.1007/s00894-012-1667-x\\nStokes, J. M., Yang, K., Swanson, K., Jin, W., Cubillos-Ruiz, A., Donghia,\\nN. M., MacNair, C. R., French, S., Carfrae, L. A., Bloom-Ackermann, Z.,Tran, V. M., Chiappino-Pepe, A., Badran, A. H., Andrews, I. W., Chory,\\nE. J., Church, J. M., Brown, E. D., Jaakkola, T. S., Barzilay, R., & Collins,\\nJ. J. (2020). A deep learning approach to antibiotic discovery. Cell,\\n180(4), 688 –702.e13. https://doi.org/10.1016/j.cell.2020.01.021\\nSu, S., Wang, Q., Xu, W., Yu, F., Hua, C., Zhu, Y., Jiang, S., & Lu, L. (2017).\\nA novel HIV-1 gp41 tripartite model for rational design of HIV-1\\nfusion inhibitors with improved antiviral activity. AIDS ,31(7), 885 –894.\\nhttps://doi.org/10.1097/QAD.0000000000001415\\nSulimov, A. V., Kutov, D. C., Katkova, E. V., & Sulimov, V. B. (2017).\\nCombined docking with classical force field and quantum chemical\\nsemiempirical method PM7. Advances in Bioinformatics ,2017 ,\\n7167691. https://doi.org/10.1155/2017/7167691\\nSun, H., Li, Y., Tian, S., Xu, L., & Hou, T. (2014). Assessing the performance\\nof MM/PBSA and MM/GBSA methods. 4. Accuracies of MM/PBSA and\\nMM/GBSA methodologies evaluated by various simulation protocols\\nusing PDBbind data set. Physical Chemistry Chemical Physics ,16(31),\\n16719 –16729. https://doi.org/10.1039/c4cp01388c\\nTanimoto, T. T. (1957). IBM Internal Report 17th Nov.; 1957.\\nThirumurugan, P., Matosiuk, D., & Jozwiak, K. (2013). Click chemistry for\\ndrug development and diverse chemical-biology applications.Chemical Reviews ,\\n113(7), 4905 –4979. https://doi.org/10.1021/\\ncr200409f\\nVamathevan, J., Clark, D., Czodrowski, P., Dunham, I., Ferran, E., Lee, G.,\\nLi, B., Madabhushi, A., Shah, P., Spitzer, M., & Zhao, S. (2019).Applications of machine learning in drug discovery and development.Nature Reviews. Drug Discovery ,18(6), 463 –477. https://doi.org/10.\\n1038/s41573-019-0024-5\\nVan der Maaten, L., & Hinton, J. (2008). Visualizing data using t-SNE.\\nJournal of Machine Learning Research ,9, 2579 –2605.\\nVerma, J., Khedkar, V. M., & Coutinho, E. C. (2010). 3D-QSAR in drug\\ndesign –A review. Current Topics in Medicinal Chemistry ,10(1),\\n95–115. https://doi.org/10.2174/156802610790232260\\nWadood, A., Ahmed, N., Shah, L., Ahmad, A., Hassan, H., & Shams, S.\\n(2013). In-silico drug design: An approach which revolutionarised thedrug discovery process. OA Drug Design & Delivery ,1(1), 3 –7.7572 A. M. ANDRIANOV ET AL.Wang, J., Wolf, R. M., Caldwell, J. W., Kollman, P. A., & Case, D. A. (2004).\\nDevelopment and testing of a general Amber force field. Journal of\\nComputational Chemistry ,25(9), 1157 –1174. https://doi.org/10.1002/\\njcc.20035\\nWeiss, M. S., Brandl, M., S €uhnel, J., Pal, D., & Hilgenfeld, R. (2001). More\\nhydrogen bonds for the (structural) biologist. Trends in Biochemical\\nSciences , 26(9), 521 –523. https://doi.org/10.1016/S0968-0004\\n(01)01935-1\\nWilen, C. B., Tilton, J. C., & Doms, R. W. (2012). HIV: Cell binding and\\nentry. Cold Spring Harbor Perspectives in Medicine ,2(8), a006866.\\nhttps://doi.org/10.1101/cshperspect.a006866\\nW/C19ojcikowski, M., Ballester, P., & Siedlecki, P. (2017). Performance of\\nmachine-learning scoring functions in structure-based virtual screen-ing. Scientific Reports ,7, 46710. https://doi.org/10.1038/srep46710\\nXiong, G.-L., Ye, W.-L., Shen, C., Lu, A.-P., Hou, T.-J., & Cao, D.-S. (2020).\\nImproving structure-based virtual screening performance via learning\\nfrom scoring function components. Briefings in Bioinformatics ,\\nbbaa094. https://doi.org/10.1093/bib/bbaa094\\nXu, B., Wang, N., Chen, T., & Li, M. (2015). Empirical evaluation of recti-\\nfied activations in convolutional network. ArXiv: preprint arXiv:\\n1505.00853.\\nXu, L., Sun, H., Li, Y., Wang, J., & Hou, T. (2013). Assessing the perform-\\nance of MM/PBSA and MM/GBSA methods. 3. The impact of forcefields and ligand charge models. The Journal of Physical Chemistry. B ,\\n117(28), 8408 –8421. https://doi.org/10.1021/jp404160y\\nYang, X., Wang, Y., Byrne, R., Schneider, G., & Yang, S. (2019). Concepts\\nof artificial intelligence for computer-assisted drug discovery.\\nChemical Reviews ,119(18), 10520 –10594. https://doi.org/10.1021/acs.\\nchemrev.8b00728\\nYilmazer, N. D., & Korth, M. (2016). Prospects of applying enhanced semi-\\nempirical QM methods for 2101 virtual drug design. Current Medicinal\\nChemistry , 23(20), 2101 –2111. https://doi.org/10.2174/\\n0929867323666160517120005\\nYu, L., Zhang, W., Wang, J., & Yu, Y. (2016). SeqGAN: Sequence genera-\\ntive adversarial nets with policy gradient. arXiv:1609.05473.\\nZhang, J., Mercado, R., Engkvist, O., & Chen, H. (2020). Comparative study\\nof deep generative models on chemical space coverage. ChemRxiv.\\nPreprint. https://doi.org/10.26434/chemrxiv.13234289.v1\\nZ h a v o r o n k o v ,A . ,I v a n e n k o v ,Y .A . ,A l i p e r ,A . ,V e s e l o v ,M .S . ,A l a d i n s k i y ,\\nV. A., Aladinskaya, A. V., Terentiev, V. A., Polykovskiy, D. A.,\\nKuznetsov, M. D., Asadulaev, A., Volkov, Y., Zholus, A.,Shayakhmetov, R. R., Zhebrak, A., Minaeva, L. I., Zagribelnyy, B. A.,\\nLee, L. H., Soll, R., Madge, D., …Aspuru-Guzik, A. (2019). Deep\\nlearning enables rapid identification of potent DDR1 kinase inhibi-\\ntors. Nature Biotechnology ,37(9), 1038 –1040.\\nhttps://doi.org/10.1038/\\ns41587-019-0224-xJOURNAL OF BIOMOLECULAR STRUCTURE AND DYNAMICS 7573'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 29
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5S0GgIQs4Rps"
      },
      "source": [
        "Initialize the connection to your database:\n",
        "\n",
        "_(do not worry if you see a few warnings, it's just that the drivers are chatty about negotiating protocol versions with the DB.)_"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 30,
      "metadata": {
        "id": "zFBR5HnZSPmK",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "f923fb15-0dc3-4840-c003-55e04d5a6cee"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:cassandra.cluster:Downgrading core protocol version from 66 to 65 for 5d9400f9-7c66-49dc-b310-a4e59c3ccb2b-us-east1.db.astra.datastax.com:29042:98602eb1-d332-4e66-b7c0-b2dfafe32ce7. To avoid this, it is best practice to explicitly set Cluster(protocol_version) to the version supported by your cluster. http://datastax.github.io/python-driver/api/cassandra/cluster.html#cassandra.cluster.Cluster.protocol_version\n",
            "WARNING:cassandra.cluster:Downgrading core protocol version from 65 to 5 for 5d9400f9-7c66-49dc-b310-a4e59c3ccb2b-us-east1.db.astra.datastax.com:29042:98602eb1-d332-4e66-b7c0-b2dfafe32ce7. To avoid this, it is best practice to explicitly set Cluster(protocol_version) to the version supported by your cluster. http://datastax.github.io/python-driver/api/cassandra/cluster.html#cassandra.cluster.Cluster.protocol_version\n",
            "ERROR:cassandra.connection:Closing connection <LibevConnection(132014064673664) 5d9400f9-7c66-49dc-b310-a4e59c3ccb2b-us-east1.db.astra.datastax.com:29042:98602eb1-d332-4e66-b7c0-b2dfafe32ce7> due to protocol error: Error from server: code=000a [Protocol error] message=\"Beta version of the protocol used (5/v5-beta), but USE_BETA flag is unset\"\n",
            "WARNING:cassandra.cluster:Downgrading core protocol version from 5 to 4 for 5d9400f9-7c66-49dc-b310-a4e59c3ccb2b-us-east1.db.astra.datastax.com:29042:98602eb1-d332-4e66-b7c0-b2dfafe32ce7. To avoid this, it is best practice to explicitly set Cluster(protocol_version) to the version supported by your cluster. http://datastax.github.io/python-driver/api/cassandra/cluster.html#cassandra.cluster.Cluster.protocol_version\n"
          ]
        }
      ],
      "source": [
        "cassio.init(token=ASTRA_DB_APPLICATION_TOKEN, database_id=ASTRA_DB_ID)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ex7NxZYb4Rps"
      },
      "source": [
        "Create the LangChain embedding and LLM objects for later usage:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 31,
      "metadata": {
        "id": "TavS0AK2SLrL"
      },
      "outputs": [],
      "source": [
        "llm = OpenAI(openai_api_key=OPENAI_API_KEY)\n",
        "embedding = OpenAIEmbeddings(openai_api_key=OPENAI_API_KEY)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9HMMx5Pm4Rpt"
      },
      "source": [
        "Create your LangChain vector store ... backed by Astra DB!"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 32,
      "metadata": {
        "id": "bg9VAk4USQvU"
      },
      "outputs": [],
      "source": [
        "astra_vector_store = Cassandra(\n",
        "    embedding=embedding,\n",
        "    table_name=\"qa_mini_demo\",\n",
        "    session=None,\n",
        "    keyspace=None,\n",
        ")"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from langchain.text_splitter import CharacterTextSplitter\n",
        "# We need to split the text using Character Text Split such that it sshould not increse token size\n",
        "text_splitter = CharacterTextSplitter(\n",
        "    separator = \"\\n\",\n",
        "    chunk_size = 800,\n",
        "    chunk_overlap  = 200,\n",
        "    length_function = len,\n",
        ")\n",
        "texts = text_splitter.split_text(raw_text)"
      ],
      "metadata": {
        "id": "9FMAhKr77AVO"
      },
      "execution_count": 33,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "texts[:50]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "k8BDHAyT7Gjr",
        "outputId": "7ab4a064-8e03-4419-db19-309ba051ae26"
      },
      "execution_count": 34,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['Full Terms & Conditions of access and use can be found at\\nhttps://www.tandfonline.com/action/journalInformation?journalCode=tbsd20\\nJournal of Biomolecular Structure and Dynamics\\nISSN: (Print) (Online) Journal homepage: www.tandfonline.com/journals/tbsd20\\nApplication of deep learning and molecular\\nmodeling to identify small drug-like compounds as\\npotential HIV-1 entry inhibitors\\nAlexander M. Andrianov, Grigory I. Nikolaev, Nikita A. Shuldov, Ivan P. Bosko,\\nArseny I. Anischenko & Alexander V. Tuzikov\\nTo cite this article:  Alexander M. Andrianov, Grigory I. Nikolaev, Nikita A. Shuldov, Ivan P.\\nBosko, Arseny I. Anischenko & Alexander V. Tuzikov (2022) Application of deep learning\\nand molecular modeling to identify small drug-like compounds as potential HIV-1 entry',\n",
              " 'Bosko, Arseny I. Anischenko & Alexander V. Tuzikov (2022) Application of deep learning\\nand molecular modeling to identify small drug-like compounds as potential HIV-1 entry\\ninhibitors, Journal of Biomolecular Structure and Dynamics, 40:16, 7555-7573, DOI:\\n10.1080/07391102.2021.1905559\\nTo link to this article:  https://doi.org/10.1080/07391102.2021.1905559\\nPublished online: 15 Apr 2021.\\nSubmit your article to this journal \\nArticle views: 612\\nView related articles \\nView Crossmark data\\nCiting articles: 7 View citing articles \\nApplication of deep learning and molecular modeling to identify small drug-like\\ncompounds as potential HIV-1 entry inhibitors\\nAlexander M. Andrianova, Grigory I. Nikolaevb, Nikita A. Shuldovc, Ivan P. Boskob, Arseny I. Anischenkocand\\nAlexander V. Tuzikovb',\n",
              " 'compounds as potential HIV-1 entry inhibitors\\nAlexander M. Andrianova, Grigory I. Nikolaevb, Nikita A. Shuldovc, Ivan P. Boskob, Arseny I. Anischenkocand\\nAlexander V. Tuzikovb\\naInstitute of Bioorganic Chemistry, National Academy of Sciences of Belarus, Minsk, Republic of Belarus;bUnited Institute of Informatics\\nProblems, National Academy of Sciences of Belarus, Minsk, Republic of Belarus;cFaculty of Applied Mathematics & Computer Science,\\nBelarusian State University, Minsk, Republic of Belarus\\nCommunicated by Ramaswamy H. Sarma\\nABSTRACT\\nA generative adversarial autoencoder for the rational design of potential HIV-1 entry inhibitors able to\\nblock CD4-binding site of the viral envelope protein gp120 was developed. To do this, the following',\n",
              " 'block CD4-binding site of the viral envelope protein gp120 was developed. To do this, the following\\nstudies were carried out: (i) an autoencoder architecture was constructed; (ii) a virtual compoundlibrary of potential anti-HIV-1 agents for training the neural network was formed by the concept of\\nclick chemistry allowing one to generate a large number of drug candidates by their assembly from\\nsmall modular units; (iii) molecular docking of all compounds from this library with gp120 was made\\nand calculations of the values of binding free energy were performed; (iv) molecular fingerprints of\\nchemical compounds from the training dataset were generated; (v) training of the developed autoen-coder was implemented followed by the validation of this neural network using more than 21 million',\n",
              " 'chemical compounds from the training dataset were generated; (v) training of the developed autoen-coder was implemented followed by the validation of this neural network using more than 21 million\\nmolecules from the ZINC15 database. As a result, three small drug-like compounds that exhibited the\\nhigh-affinity binding to gp120 were identified. According to the data from molecular docking, machine\\nlearning, quantum chemical calculations, and molecular dynamics simulations, these compounds show\\nthe low values of binding free energy in the complexes with gp120 similar to those calculated usingthe same computational protocols for the HIV-1 entry inhibitors NBD-11021 and NBD-14010, highly\\npotent and broad anti-HIV-1 agents presenting a new generation of the viral CD4 antagonists. The',\n",
              " 'potent and broad anti-HIV-1 agents presenting a new generation of the viral CD4 antagonists. The\\nidentified CD4-mimetic candidates are suggested to present good scaffolds for the design of novelantiviral drugs inhibiting the early stages of HIV-1 infection.ARTICLE HISTORY\\nReceived 5 September 2020\\nAccepted 26 February 2021\\nKEYWORDS\\nHIV-1; gp120; HIV-1 entryinhibitors; deep learning;generative adversarial\\nautoencoder; virtual\\nscreening; moleculardocking; quantum chemicalcalculations; molecular\\ndynamics simulations;\\nbinding free energycalculations;anti-HIV-1 drugs\\nIntroduction\\nModern methods of computer-aided drug design significantly',\n",
              " 'dynamics simulations;\\nbinding free energycalculations;anti-HIV-1 drugs\\nIntroduction\\nModern methods of computer-aided drug design significantly\\nexpand the capabilities of the pharmaceutical industry, vastlyreducing the time and costs required for developing noveltherapeutic agents (Katsila et al., 2016 ; Knight-Schrijver et al.,\\n2016 ; Leelananda & Lindert, 2016 ; Mallipeddi et al., 2014 ;\\nSliwoski et al., 2014 ; Wadood et al., 2013 ). Despite the effi-\\ncacy of computational methods in drug design is currently',\n",
              " '2016 ; Leelananda & Lindert, 2016 ; Mallipeddi et al., 2014 ;\\nSliwoski et al., 2014 ; Wadood et al., 2013 ). Despite the effi-\\ncacy of computational methods in drug design is currently\\nuniversally recognized, the development of new mathemat-ical approaches and the availability of powerful and cheapcomputing resources promote their continuous improve-ment. Among these approaches, machine learning methodsand, in particular, deep learning techniques, which offersgreat potential for further progress in this research area,occupy an important place (Dobchev et al., 2014 ; Lipinski\\net al., 2019 ; Vamathevan et al., 2019 ). To date, computer-',\n",
              " 'et al., 2019 ; Vamathevan et al., 2019 ). To date, computer-\\naided design of potential drugs using machine learningmethods is one of the most important and rapidly develop-ing areas of chem- and bioinformatics (Dobchev et al., 2014 ;\\nLipinski et al., 2019 ; Vamathevan et al., 2019 ). Unlike physical\\nmodels based on explicit physical equations, such asquantum chemistry or molecular dynamics (MD) modeling,\\nmachine learning approaches use pattern recognition algo-rithms to determine the mathematical relationships betweenempirical observations of small molecules and their extrapo-\\nlation to predict chemical, biological and physical properties\\nof new compounds (Dobchev et al., 2014 ; Lipinski et al.,\\n2019 ; Vamathevan et al., 2019 ). In addition, compared to',\n",
              " 'lation to predict chemical, biological and physical properties\\nof new compounds (Dobchev et al., 2014 ; Lipinski et al.,\\n2019 ; Vamathevan et al., 2019 ). In addition, compared to\\nphysical models, machine learning methods are more effi-cient and can easily scale to big data sets. One of the bene-fits of using machine learning to design drugs is to helpresearchers understand and use the relationships between\\nchemical structures and their biological activity (Dobchev\\net al., 2014 ; Lipinski et al., 2019 ; Vamathevan et al., 2019 ).',\n",
              " 'chemical structures and their biological activity (Dobchev\\net al., 2014 ; Lipinski et al., 2019 ; Vamathevan et al., 2019 ).\\nModern machine learning methods can be used to modelthe quantitative structure-activity relationship (QSAR) orquantitative structure-property relations (QSPR) and developintelligent tools that can accurately predict the effect ofchemical modifications of a compound on its biological activ-\\nity, pharmacokinetic and toxicological characteristics\\n(Cherkasov et al., 2014 ; Verma et al., 2010 ). In addition, these\\nmethods can be applied for extraction of descriptors fromchemical structures automatically and directly (Duvenaud\\nCONTACT Alexander M. Andrianov alexande.andriano@yandex.ru Institute of Bioorganic Chemistry, National Academy of Sciences of Belarus, Minsk,',\n",
              " 'CONTACT Alexander M. Andrianov alexande.andriano@yandex.ru Institute of Bioorganic Chemistry, National Academy of Sciences of Belarus, Minsk,\\nRepublic of Belarus\\n/C2232021 Informa UK Limited, trading as Taylor & Francis GroupJOURNAL OF BIOMOLECULAR STRUCTURE AND DYNAMICS\\n2022, VOL. 40, NO. 16, 7555 –7573\\nhttps://doi.org/10.1080/07391102.2021.1905559et al., 2015 ; Jaeger et al., 2018 ; Kadurin, Nikolenko et al.,\\n2017 ), drug repurposing (Agastheeswaramoorthy &\\nSevilimedu, 2020 ; Kinnings et al., 2011 ), protein structure pre-\\ndiction (Senior et al., 2020 ), virtual screening of drug candi-\\ndates and prediction of protein/ligand binding affinity (Chenet al., 2018 ; Ekins, 2016 ; Jimenez-Luna et al., 2020 ; Lavecchia,\\n2019 ; Li et al., 2021 ; Stokes et al., 2020 ; Xiong et al., 2020 ;',\n",
              " 'dates and prediction of protein/ligand binding affinity (Chenet al., 2018 ; Ekins, 2016 ; Jimenez-Luna et al., 2020 ; Lavecchia,\\n2019 ; Li et al., 2021 ; Stokes et al., 2020 ; Xiong et al., 2020 ;\\nYang et al., 2019 ), and related aspects.\\nBesides these prediction applications, machine learning\\ncan also be used for generation of novel compounds with\\ndefined structural and functional properties (Goodfellowet al., 2014 ; Gupta et al., 2018 ; Olivecrona et al., 2017 ;\\nSanchez-Lengeling et al., 2017 ; Silver et al., 2016 ; Yu et al.,\\n2016 ; Zhang et al., 2020 ). Over the last 5 years, deep learning\\ngenerative models have emerged as novel methods for de\\nnovo drug design. Due to the tremendous progress of deep\\nlearning techniques, generative deep learning models with',\n",
              " 'generative models have emerged as novel methods for de\\nnovo drug design. Due to the tremendous progress of deep\\nlearning techniques, generative deep learning models with\\ndifferent architectures, data types, and learning methods thatintroduce an encouraging solution to drug discovery have\\nbeen currently developed (Ar /C19us-Pous, Blaschke et al., 2019 ;\\nAr/C19us-Pous, Johansson et al., 2019 ; Bian et al., 2019 ; Blaschke\\net al., 2020 ;G/C19omez-Bombarelli et al., 2018 ; Guimaraes et al.,\\n2017 ;M/C19endez-Lucio et al., 2020 ; Mercado et al., 2020 ;\\nPrykhodko et al., 2019 ; Segler et al., 2018 ; Zhavoronkov et al.,\\n2019 ). The recent applications of the generative deep learn-\\ning models clearly showed their ability to generate molecules',\n",
              " 'Prykhodko et al., 2019 ; Segler et al., 2018 ; Zhavoronkov et al.,\\n2019 ). The recent applications of the generative deep learn-\\ning models clearly showed their ability to generate molecules\\nthat can be synthesized, are active in vitro , metabolically sta-\\nble, and exhibit in vivo activity in disease-relevant models. In\\nparticular, the Janus kinase 3 inhibitor (JAK3) presenting a\\nnew class of immunomodulatory agents was newly devel-\\noped using the conditional adversarial autoencoder\\n(Polykovskiy et al., 2018 ). Furthermore, in vivo active inhibi-\\ntors of discoidin domain receptors 1 and 2 (DDR1 and DDR2)\\nwere designed using Generative Tensorial ReinforcementLearning and the pharmacokinetic profile of DDR1 was vali-\\ndated by in vivo mouse experiments (Zhavoronkov et al.,',\n",
              " 'were designed using Generative Tensorial ReinforcementLearning and the pharmacokinetic profile of DDR1 was vali-\\ndated by in vivo mouse experiments (Zhavoronkov et al.,\\n2019). In this regard, the development and application ofdeep learning generative methods for computer-aided\\ndesign of potential drugs is of great scientific and practical\\nimportance (Bian et al., 2019 ; Dobchev et al., 2014 ; Lipinski\\net al., 2019 ;M/C19endez-Lucio et al., 2020 ; Vamathevan et al.,\\n2019 ; Zhang et al., 2020 ; Zhavoronkov et al., 2019).\\nAn essential role in structure-based drug discovery\\nbelongs to molecular docking that is widely used to predict\\nthe ligand conformation and its position and orientation',\n",
              " 'An essential role in structure-based drug discovery\\nbelongs to molecular docking that is widely used to predict\\nthe ligand conformation and its position and orientation\\nwithin the binding site of target proteins, assess the bindingaffinity and investigate the interaction profile of drug candi-dates (Guedes et al., 2018 ; Meng et al., 2011 ). The latest\\ndevelopments of semiempirical quantum mechanical and\\ndensity functional theory methods as well as applications ofexplicit quantum mechanical calculations to structure-based\\ndrug design in the context of identification and optimization\\nof drug candidates show the growing importance of quan-tum chemistry in the study of protein-ligand interaction\\n(Cavasotto et al., 2018 ; Ryde & S €oderhjelm, 2016 ; Yilmazer &',\n",
              " 'of drug candidates show the growing importance of quan-tum chemistry in the study of protein-ligand interaction\\n(Cavasotto et al., 2018 ; Ryde & S €oderhjelm, 2016 ; Yilmazer &\\nKorth, 2016 ). Molecular dynamics is also a powerful approach\\nfor structure-based drug discovery (Hollingsworth & Dror,\\n2018 ). Unlike molecular docking, MD simulates moving eachatom separately in the field of the rest atoms and represents\\nthe flexibility of both the ligand and protein more effectivelythan other algorithms. Examples of successful applicationsclearly demonstrate the power of computational approachesto identify compounds with desired properties from large\\ndatabases and design novel small-molecule drug candidates\\n(Hollingsworth & Dror, 2018 ).\\nIn recent years, a large number of studies have appeared',\n",
              " 'databases and design novel small-molecule drug candidates\\n(Hollingsworth & Dror, 2018 ).\\nIn recent years, a large number of studies have appeared\\non the use of machine learning methods to predict potential\\nHIV-1 inhibitors and virus resistance to anti-HIV drugs\\n(reviewed in Dubey, 2018 ). However, all of these studies\\nfocus on viral enzymes, namely reverse transcriptase, prote-ase and integrase (Dubey, 2018 ; Qureshi et al., 2018 ).\\nCompounds that block these enzymes cannot prevent the\\nvirus from entering a target cell, which raises attention to\\nthe HIV-1 inhibitors that can interfere with the initial stagesof the HIV-1 infection cycle by blocking the viral adsorptionto CD4\\nþcells or/and the virus-cell membrane fusion\\n(Andrianov et al., 2017 ,2019 ; De Clercq, 2005 ; Este & Telenti,',\n",
              " 'þcells or/and the virus-cell membrane fusion\\n(Andrianov et al., 2017 ,2019 ; De Clercq, 2005 ; Este & Telenti,\\n2007 ; Li et al., 2017 ; Rusconi et al., 2007 ; Ryser & Fluckiger,\\n2005 ; Su et al., 2017 ).\\nAt the first stage of replication cycle, the HIV-1 envelope\\ngp120 protein binds to the primary cellular receptor CD4 and\\nthen to the chemokine co-receptors CCR5 and/or CXCR4\\n(Wilen et al., 2012 ). Sequential interactions of gp120 with CD4\\nand CCR5/CXCR4 trigger structural changes in the HIV-1 gp41transmembrane domain, resulting in the Env-mediated mem-brane fusion and viral entry into CD4 þcells (Wilen et al.,\\n2012 ). The HIV-1 entry into the host cell is therefore a highly',\n",
              " '2012 ). The HIV-1 entry into the host cell is therefore a highly\\npromising target for several types of antiviral agents, such asinhibitors of the gp120 /C0CD4 interaction, antagonists of the\\nco-receptors CCR5 and CXCR4, and inhibitors of the virus-cellmembrane fusion (Andrianov et al., 2017 ,2019 ).\\nAt present, many HIV-1 entry/fusion inhibitors have been\\ndeveloped (Andrianov et al., 2017 ,2019 ; De Clercq, 2005 ;E s t e\\n& Telenti, 2007 ;L ie ta l . , 2017 ; Rusconi et al., 2007 ;R y s e r&\\nFluckiger, 2005 ;S ue ta l . , 2017 ), but only two of these antiviral\\nagents are currently used in clinical practice for treatment of\\nthe HIV-1 infection. These are maravirok that binds to co-receptor CCR5 (MacArthur & Novak, 2008 ) and a peptide',\n",
              " 'agents are currently used in clinical practice for treatment of\\nthe HIV-1 infection. These are maravirok that binds to co-receptor CCR5 (MacArthur & Novak, 2008 ) and a peptide\\ninhibitor enfuvirtide blocking the gp41-mediated virus fusion(Matthews et al., 2004 ). However, like the anti-HIV-1 drugs\\nthat act inside the target cell, these compounds have a num-\\nber of limitations and severe side effects (Arts & Hazuda, 2012 ;\\nKumari & Singh, 2012 ), which makes it urgent to search for\\nnew potent HIV-1 entry inhibitors with acceptable ADMET(absorption, distribution, metabolism, excretion, and toxicity)\\nparameters (Antoine et al., 2017 ;K e r n s&D i , 2008 ).\\nBecause the HIV-1 gp120 binding to CD4 starts the process',\n",
              " 'parameters (Antoine et al., 2017 ;K e r n s&D i , 2008 ).\\nBecause the HIV-1 gp120 binding to CD4 starts the process\\nof viral entry into susceptible cells, one of the most promisingareas in the design of novel antiviral drugs involves the search\\nfor small-molecule compounds able to break the key interac-\\ntions between gp120 and CD4. Despite numerous reports ofnew CD4-mimetic candidates, there are currently no licensedanti-HIV-1 drugs mimicking cellular receptor CD4 (reviewed inLi et al., 2017 ). In this connection, studies on the development\\nof selective and potent drug-like compounds blocking\\ngp120 /C0CD4 interface still cause great interest.7556 A. M. ANDRIANOV ET AL.The goal of this work was to develop and apply a genera-',\n",
              " 'of selective and potent drug-like compounds blocking\\ngp120 /C0CD4 interface still cause great interest.7556 A. M. ANDRIANOV ET AL.The goal of this work was to develop and apply a genera-\\ntive adversarial autoencoder to identify potential HIV-1 entryinhibitors that target the hydrophobic Phe-43 cavity ofgp120 critically important for the virus attachment to theCDþT cells (Kwong et al., 1998 ).\\nIn doing so, the following studies were carried out: (i)\\narchitecture of the generative adversarial autoencoder wasconstructed; (ii) a virtual compound library of potential anti-HIV-1 agents for training the neural network was formed by\\nthe concept of click chemistry allowing one to generate a',\n",
              " 'the concept of click chemistry allowing one to generate a\\nlarge number of drug candidates by their assembly fromsmall modular units (Hein et al., 2008 ; Kolb et al., 2001 ;\\nMoses & Moorhouse, 2007 , p. 32; Thirumurugan et al., 2013 );\\n(iii) molecular docking of all compounds from this librarywith gp120 was made and calculations of the values of bind-\\ning free energy were performed; (iv) molecular fingerprints\\nfor chemical compounds from the training dataset were gen-erated; (v) training the neural network was implemented fol-lowed by validation of the learning outcomes and work ofthe autoencoder; (vi) methods of molecular docking were\\nused to select the most promising anti-HIV-1 drug candidate',\n",
              " 'used to select the most promising anti-HIV-1 drug candidate\\nfrom the ensemble of compounds, which were identified bythe fingerprint-based screening of the virtual compounddatabase when the autoencoder validation; (vii) the ligand-binding affinity of the identified drug candidate to gp120\\nwas predicted in terms of the scoring functions of molecular\\ndocking, machine learning, semiempirical quantum chemis-try, and molecular dynamics; (viii) the values of bindingenergy approximated by these scoring functions were com-pared to those predicted using the same computational pro-tocols for the HIV-1 inhibitors NBD-11021 and NBD-14010,\\nhighly potent and broad anti-HIV-1 agents presenting a new\\ngeneration of viral entry antagonists (Acharya et al., 2015 ;\\nCurreli et al., 2015 ,2017 , 2018, 2020).',\n",
              " 'highly potent and broad anti-HIV-1 agents presenting a new\\ngeneration of viral entry antagonists (Acharya et al., 2015 ;\\nCurreli et al., 2015 ,2017 , 2018, 2020).\\nMethods and materials\\nDevelopment of a generative adversarial autoencoder\\nAs the most commonly used machine learning models oper-\\nate with a fixed input size, the generative network beingdeveloped needed to convert a huge variety of molecular\\nstructures into the form of data appropriate for the autoen-\\ncoder under development. For this reason, the MACCS fin-gerprints ( http://www.dalkescientific.com/writings/NBN/finger\\nprints.html ;https://www.rdkit.org/ ) were chosen in this study\\nas the type of descriptors to train the model, given the suc-\\ncessful example of applying adversarial autoencoder for gen-',\n",
              " 'prints.html ;https://www.rdkit.org/ ) were chosen in this study\\nas the type of descriptors to train the model, given the suc-\\ncessful example of applying adversarial autoencoder for gen-\\nerating fingerprints encoding structures of potentialanticancer agents (Kadurin, Aliper et al., 2017 ). That is why\\nthis neural network was used in our study as a basic genera-tive model to solve the task.\\nFigure 1 sheds light on the architecture of the developed\\nadversarial autoencoder. The model consists of two neuralnetworks including an autoencoder and a discriminator thatwork during training in an adversarial mode ( Figure 1 ). This\\nallows one to configure the autoencoder parameters in the\\ntraining mode and provide obtaining a high-quality output',\n",
              " 'allows one to configure the autoencoder parameters in the\\ntraining mode and provide obtaining a high-quality output\\ndata at the next stage of their generation. The task of thediscriminator is to distinguish real data from those generated\\nby the encoder. The developed autoencoder is a seven-layer\\nneural network with input and output layers, a latent layer,and four fully connected layers ( Figure 1 ).\\nFingerprints of chemical compounds are fed to the input\\nlayer, the data of which pass through two fully connected\\nlayers (encoder) and fall on the latent layer where a numer-\\nical estimate of the binding energy to the molecular target isadded to the result obtained. Next, the fingerprints pass\\nthrough two fully connected layers (decoder) and go to the',\n",
              " 'ical estimate of the binding energy to the molecular target isadded to the result obtained. Next, the fingerprints pass\\nthrough two fully connected layers (decoder) and go to the\\noutput, which, like the input, is a fingerprint vector. Thelatent layer consists of three neurons two of which receive\\nvalues from the encoder, and the third one receives the\\nvalue of the binding free energy. In the autoencoder genera-tive mode, random numbers are entered to the two neurons\\nof the latent layer and a binding free energy value to the\\nthird neuron, and then pass through the decoder generatingfingerprints of molecules satisfying a pre-defined threshold\\nenergy value. For the generation of such molecules, it is',\n",
              " 'third neuron, and then pass through the decoder generatingfingerprints of molecules satisfying a pre-defined threshold\\nenergy value. For the generation of such molecules, it is\\nimportant that the data entering the latent layer after pass-ing through the encoder have a normal distribution. To\\nensure this condition, in the course of adversarial training of\\nthe encoder and discriminator, it was aimed that the encoderwas able to produce data with a normal distribution on thelatent layer, and that the discriminator to distinguish the\\nstandard normal distribution (generated data) from the distri-\\nbution obtained on the latent layer.\\nAs noted before, the developed adversarial autoencoder is\\nbased on the model of the neural network that wasdesigned to generate chemical compounds with anticancer',\n",
              " 'bution obtained on the latent layer.\\nAs noted before, the developed adversarial autoencoder is\\nbased on the model of the neural network that wasdesigned to generate chemical compounds with anticancer\\nproperties (Kadurin, Aliper et al., 2017 ), but has the following\\nspecific features ( Figure 1 ):\\n/C15Instead of a neuron liable for percent growth inhibition\\nof cells (Kadurin, Aliper et al., 2017 ), a neuron responsible\\nfor the values of binding free energy is added to the\\nlatent layer. This neuron does not interact with theencoder and is fed only to the input of the decoder\\ntogether with the data obtained by the encoder. At the\\nlearning stage, the values of this neuron are assignedaccording to the estimation of value of binding free\\nenergy for the testing compounds. At the generative',\n",
              " 'learning stage, the values of this neuron are assignedaccording to the estimation of value of binding free\\nenergy for the testing compounds. At the generative\\nstage, its values have the meaning of a binding freeenergy threshold. The latent layer contains three neurons;\\n/C15The encoder consists of two consecutive layers L1 and L2\\nwith 32 and 16 neurons, respectively. The decoder\\nincludes two layers L ’1 and L ’2 containing 16 and 32 neu-\\nrons, respectively;\\n/C15The discriminator consists of four layers including 2, 16, 3\\nand 1 neurons, respectively;\\n/C15The activation function leaky relu (Xu et al., 2015 ) is used\\non the intermediate layers of the autoencoder;\\n/C15Sigmoid activation functions are used on all layers of the\\ndiscriminator (Gibbs & MacKay, 2000 );',\n",
              " 'on the intermediate layers of the autoencoder;\\n/C15Sigmoid activation functions are used on all layers of the\\ndiscriminator (Gibbs & MacKay, 2000 );\\n/C15To provide an additional level of protection against over-\\nfitting, a dropout layer is used between two fully con-nected layers of the encoder and decoder, allowing theJOURNAL OF BIOMOLECULAR STRUCTURE AND DYNAMICS 7557neural network to completely use its parameters (weights)\\nand selective random disconnection during the training.\\nA three-stage iterative procedure was used for training\\nthe developed adversarial autoencoder, namely: (i) trainingthe discriminator to distinguish a given normal distribution\\nfrom the encoded one obtained by the encoder on the',\n",
              " 'the developed adversarial autoencoder, namely: (i) trainingthe discriminator to distinguish a given normal distribution\\nfrom the encoded one obtained by the encoder on the\\nlatent layer; (ii) joint training of the encoder and decoder asan autoencoder; (iii) training the encoder to compress the\\ndata in such a way that they look like a normal distribution.\\nDiscriminatory models are characterized by a simpler\\nlearning process than generative neural networks. Therefore,\\nduring the first attempts to train the model, a situation\\noccurred when the function of the discriminator lossesdecreased, and the function of the encoder losses grew\\n(Figure 2 ).\\nIt should be noted that when training these two networks\\nin an adversarial mode, it is difficult to determine whether',\n",
              " '(Figure 2 ).\\nIt should be noted that when training these two networks\\nin an adversarial mode, it is difficult to determine whether\\nthe encoder is really poorly trained or it can not ‘trick ’the\\ndiscriminator. Small discriminator losses combined with bigencoder losses do not necessarily indicate deficiencies in the\\ntraining of the encoder. Nevertheless, to improve the learn-\\ning process, the following measures have been taken:/C15Preliminary training of the discriminator before training\\nthe entire adversarial autoencoder. Preliminary training\\nwas carried out under the assumption that the discrimin-ator would distinguish random numbers obtained from\\nthe untrained encoder from those corresponding to the\\nnormal distribution used;\\n/C15When training the entire network, the discriminator was',\n",
              " 'the untrained encoder from those corresponding to the\\nnormal distribution used;\\n/C15When training the entire network, the discriminator was\\ntrained not every epoch, but once in 2 epochs, and the\\nautoencoder was trained every epoch;\\n/C15For the discriminator, a lower learning rate (0.001) was\\nset, while for the entire adversarial autoencoder it was\\nequal to 0.005;\\n/C15A small noise was added to the data generated from the\\nnormal distribution, which made the discriminator oper-\\nation more difficult.\\nThese changes concerned each epoch of training which\\npresented an iterative process including the following steps:\\nStep 1 /C0the encoder and decoder were trained jointly as\\nthe autoencoder (Creswell et al., 2018 );\\nStep 2 /C0the discriminator was trained to distinguish the',\n",
              " 'Step 1 /C0the encoder and decoder were trained jointly as\\nthe autoencoder (Creswell et al., 2018 );\\nStep 2 /C0the discriminator was trained to distinguish the\\ngiven normal distribution from the encoded representation\\nobtained by the encoder on the latent layer;\\nFigure 1. The architecture of a generative adversarial autoencoder for prediction of potential anti-HIV-1 agents targeting the viral envelope gp120 protein . The\\nencoder consists of two consecutive layers L1 and L2 with 32 and 16 neurons, respectively. The decoder includes two layers L ’1 and L ’2 containing 16 and 32 neu-\\nrons, respectively. The latent layer consists of 3 neurons.7558 A. M. ANDRIANOV ET AL.Step 3 /C0the encoder was trained to compress the data in\\nsuch a way that they represented a normal distribution.',\n",
              " 'rons, respectively. The latent layer consists of 3 neurons.7558 A. M. ANDRIANOV ET AL.Step 3 /C0the encoder was trained to compress the data in\\nsuch a way that they represented a normal distribution.\\nFor the above steps, the following loss functions\\nwere used:\\nStep 1 /C0The reconstruction loss function was presented\\nby a mean squared error between original fingerprints and\\nvalues decoded by the decoder part of the model. This func-\\ntion penalized the network for deviations of restored finger-prints from the input ones and was calculated using the\\nformula\\nRl¼1\\nnXn\\ni¼1ðyi/C0y0\\niÞ2,\\nwhere n is the number of bits in the fingerprint, y iis the\\nvalue at the i-th position of the original fingerprint, y ’iis the\\nvalue at the i-th position of the fingerprint restored after',\n",
              " 'iÞ2,\\nwhere n is the number of bits in the fingerprint, y iis the\\nvalue at the i-th position of the original fingerprint, y ’iis the\\nvalue at the i-th position of the fingerprint restored after\\ndecoding the values compressed by the encoder.\\nStep 2 /C0The discriminator loss function consisted of two\\nBinary Cross Entropy losses, where Binary Cross Entropy loss\\nwas (Dumoulin et al., 2017 ):\\nBCE label ,y ðÞ ¼/C0 ð label/C3logyðÞ/C01/C0label ðÞ /C3logð1/C0yÞÞ,\\nwhere label 2f0,1g–classes, and y is the discriminator ’s out-\\nput, predicted class probability.\\nThe first binary cross entropy loss related to discrimin-\\nator positive loss (the ability to correctly identify data gen-\\nerated from real normal distribution) and took as\\narguments ones and discriminator ’s prediction for values',\n",
              " 'ator positive loss (the ability to correctly identify data gen-\\nerated from real normal distribution) and took as\\narguments ones and discriminator ’s prediction for values\\nsampled from real distribution. The second binary cross\\nentropy loss related to discriminator negative loss (the abil-\\nity to distinguish encoded data from real distribution). Ittook as arguments zeros and discriminator ’s output for\\nencoded values.\\nThe total equation used for calculation of the discrimin-\\nator loss was\\nDLF¼BCEð1,DðNDÞÞ þ BCEð0,DðEDÞÞ,\\nwhere D(ND) is the discriminator predicted value for the nor-\\nmal distribution, and D(ED) is the discriminator output for\\nthe encoded data (Janocha & Czarnecki, 2017 ).\\nStep 3 /C0The encoder loss function exposed the ability of',\n",
              " 'mal distribution, and D(ED) is the discriminator output for\\nthe encoded data (Janocha & Czarnecki, 2017 ).\\nStep 3 /C0The encoder loss function exposed the ability of\\nencoder to deceive discriminator and it was represented by\\na binary cross entropy loss between ones and encoded val-\\nues that were passed through discriminator. This functionwas calculated using the formulaELF¼BCEð1,DðEDÞÞ:\\nThe discriminator and autoencoder were trained jointly in\\ntwo stages: reconstruction stage (step 1) and regularization',\n",
              " 'The discriminator and autoencoder were trained jointly in\\ntwo stages: reconstruction stage (step 1) and regularization\\nstage (steps 2 and 3) which were performed in each subsetfrom the original data. At the reconstruction stage, theautoencoder updated the encoder and decoder to minimizethe recovery error of the input and output data. At the regu-larization stage, the discriminator network was first updated(step 2) to distinguish the true samples obtained by the nor-mal distribution generator from the compressed input dataon the latent layer calculated by the autoencoder, and then(step 3) the autoencoder updated its encoder to confuse thediscriminator network.\\nFigure 3 shows the graphs of the loss functions taking',\n",
              " 'Figure 3 shows the graphs of the loss functions taking\\ninto account the changes made above. The discriminator lossfunction graph was further divided into two graphs present-ing the loss function of the correct classification of the nor-mal distribution and the loss function for the encoded data.As follows from Figure 3 , the discriminator was well trained\\nto classify data from a normal distribution, but its ability tocorrectly identify the data received by theencoder decreased.\\nThe following parameters were used to train the\\nautoencoder:\\n/C15The number of epochs for the main version of the model\\nused for generation was 400.\\n/C15The learning rate of the entire autoencoder in step 1\\nwas 0.005.\\n/C15The learning rate of the discriminator in step 2 was 0.001.',\n",
              " 'used for generation was 400.\\n/C15The learning rate of the entire autoencoder in step 1\\nwas 0.005.\\n/C15The learning rate of the discriminator in step 2 was 0.001.\\n/C15The encoder learning rate at step 3 was 0.005.\\n/C15The Batch size parameter was set to 128.\\n/C15Optimizer used the Adam method (Kingma & Ba, 2014 ).\\nWhen training the neural network, it was important to',\n",
              " '/C15The Batch size parameter was set to 128.\\n/C15Optimizer used the Adam method (Kingma & Ba, 2014 ).\\nWhen training the neural network, it was important to\\nmake sure that in the generation mode the model is able tomake different results for various input data and producemany molecular fingerprints. At the same time, this modelwas not supposed to generate fingerprints from a smallnumber of possible options, since there was a probabilitythat such a situation reflected some minimum of the initialloss functions. In order to reduce the probability of such ascenario, the tSNE algorithm (Van der Maaten & Hinton,2008 ) was used to subsample the generated molecular',\n",
              " 'Figure 2. The discriminator and encoder loss functions with faster discriminator training. The graphs show the informative part of the functions.JOURNAL OF BIOMOLECULAR STRUCTURE AND DYNAMICS 7559fingerprints ( Figure 4 ). Each generated fingerprint represents\\nmolecular descriptor as a two- or three-dimensional point in\\nsuch a way that similar molecular descriptors are displayed\\nby closely spaced points, and dissimilar descriptors are dis-\\nplayed with a high probability by points that are far from\\neach other.\\nIt should be noted that, due to the adversarial process,\\nthe autoencoder was learned to compress the input data in\\nsuch a way that it enabled similar compounds to have\\nresembling latent vector representations, providing thereby\\ngeneration of meaningful samples in the generation mode.',\n",
              " 'such a way that it enabled similar compounds to have\\nresembling latent vector representations, providing thereby\\ngeneration of meaningful samples in the generation mode.\\nGeneration of the training dataset\\nThe training dataset was generated by the AutoClickChem\\nsoftware package (Durrant & McCammon, 2012 ). This pro-\\ngram implements the click chemistry methodology in silicoto design a large set of drug candidates that include the\\nmoieties providing the ligand with biological activity towards\\na target protein. In the first step, two compound libraries\\ncontaining small molecules with the functional groupsinvolved in the click-reaction of azide-alkyne cycloaddition\\n(Kolb et al., 2001 ) were created by virtual screening of the',\n",
              " 'containing small molecules with the functional groupsinvolved in the click-reaction of azide-alkyne cycloaddition\\n(Kolb et al., 2001 ) were created by virtual screening of the\\nDrug-Like subset of the ZINC15 database, a virtual collectionof 1.5 billion molecules designed for in silico screening\\n(Sterling & Irwin, 2015 ). For this purpose, the DataWarrior\\nprogram (Sander et al., 2015 ) was used. Aromatic compounds\\nwith molecular mass <250 Da containing azide or alkyne\\ngroups were selected and placed into library 1, and all low-\\nmolecular compounds (molecular mass <250 Da) with the\\nabove functional groups were collected in library 2. The\\nselection of aromatic compounds as a scaffold for the design',\n",
              " 'molecular compounds (molecular mass <250 Da) with the\\nabove functional groups were collected in library 2. The\\nselection of aromatic compounds as a scaffold for the design\\nof potential CD4-mimetic candidates is explained by the factthat their aromatic moieties can mimic the key interactions\\nof the Phe-43 residue of CD4 with the hydrophobic Phe-43\\nFigure 3. The discriminator loss function (DLF) consisting of the DLF for normal distribution and the DLF for encoded data, and the encoder loss function. The\\ngraphs show the informative part of the functions.\\nFigure 4. The results of the tSNE algorithm action for three (a) and two (b) components with a subsample from the generated molecular fingerprints.7560 A. M. ANDRIANOV ET AL.cavity of gp120. As follows from the crystal gp120 /C0CD4']"
            ]
          },
          "metadata": {},
          "execution_count": 34
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "V1WK54-74Rpt"
      },
      "source": [
        "### Load the dataset into the vector store\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 35,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "GX5BECsdSUUM",
        "outputId": "7d193561-5e33-4858-b6ff-d9dd3e42719f"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Inserted 50 headlines.\n"
          ]
        }
      ],
      "source": [
        "\n",
        "astra_vector_store.add_texts(texts[:50])\n",
        "\n",
        "print(\"Inserted %i headlines.\" % len(texts[:50]))\n",
        "\n",
        "astra_vector_index = VectorStoreIndexWrapper(vectorstore=astra_vector_store)"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "KhVf0kir2Uke"
      },
      "execution_count": 35,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "oLJp8yPF4Rpt"
      },
      "source": [
        "### Run the QA cycle\n",
        "\n",
        "Simply run the cells and ask a question -- or `quit` to stop. (you can also stop execution with the \"▪\" button on the top toolbar)\n",
        "\n",
        "Here are some suggested questions:\n",
        "- _What is the current GDP?_\n",
        "- _How much the agriculture target will be increased to and what the focus will be_\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 36,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MbJugrh7SX3C",
        "outputId": "cf7745b7-7f99-489e-aeb2-719ecb3c4493"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Enter your question (or type 'quit' to exit): what is the title \n",
            "\n",
            "QUESTION: \"what is the title\"\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:cassandra.protocol:Server warning: Top-K queries can only be run with consistency level ONE / LOCAL_ONE / NODE_LOCAL. Consistency level LOCAL_QUORUM was requested. Downgrading the consistency level to LOCAL_ONE.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "ANSWER: \"The title is \"Application of deep learning and molecular modeling to identify small drug-like compounds as potential HIV-1 entry inhibitors\".\"\n",
            "\n",
            "\n",
            "What's your next question (or type 'quit' to exit): when did it publish\n",
            "\n",
            "QUESTION: \"when did it publish\"\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:cassandra.protocol:Server warning: Top-K queries can only be run with consistency level ONE / LOCAL_ONE / NODE_LOCAL. Consistency level LOCAL_QUORUM was requested. Downgrading the consistency level to LOCAL_ONE.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "ANSWER: \"The article was published in 2022.\"\n",
            "\n",
            "\n",
            "What's your next question (or type 'quit' to exit): Who is the author\n",
            "\n",
            "QUESTION: \"Who is the author\"\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:cassandra.protocol:Server warning: Top-K queries can only be run with consistency level ONE / LOCAL_ONE / NODE_LOCAL. Consistency level LOCAL_QUORUM was requested. Downgrading the consistency level to LOCAL_ONE.\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ANSWER: \"The authors of this article are Alexander M. Andrianov, Grigory I. Nikolaev, Nikita A. Shuldov, Ivan P. Bosko, Arseny I. Anischenko, and Alexander V. Tuzikov.\"\n",
            "\n",
            "\n",
            "What's your next question (or type 'quit' to exit): quit\n"
          ]
        }
      ],
      "source": [
        "first_question = True\n",
        "while True:\n",
        "    if first_question:\n",
        "        query_text = input(\"\\nEnter your question (or type 'quit' to exit): \").strip()\n",
        "    else:\n",
        "        query_text = input(\"\\nWhat's your next question (or type 'quit' to exit): \").strip()\n",
        "\n",
        "    if query_text.lower() == \"quit\":\n",
        "        break\n",
        "\n",
        "    if query_text == \"\":\n",
        "        continue\n",
        "\n",
        "    first_question = False\n",
        "\n",
        "    print(\"\\nQUESTION: \\\"%s\\\"\" % query_text)\n",
        "    answer = astra_vector_index.query(query_text, llm=llm).strip()\n",
        "    print(\"ANSWER: \\\"%s\\\"\\n\" % answer)\n",
        "\n",
        "    # print(\"FIRST DOCUMENTS BY RELEVANCE:\")\n",
        "    # for doc, score in astra_vector_store.similarity_search_with_score(query_text, k=4):\n",
        "    #     print(\"    [%0.4f] \\\"%s ...\\\"\" % (score, doc.page_content[:20]))"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "dSaUPguw389l"
      },
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.12"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}